---
title: "01_pseudobulk_dea"
author: "Tabea M. Soelter"
date: '2023-09-27'
output: html_document
---
**Pseudo-bulk differential gene expression analysis using DESeq2**

__Goal__: For both datasets, I am pseudo-bulking counts by cell type in order to perform a differential gene expression analysis using DESeq2 between AD and control. This is a step needed for downstream differential biological activity analyses.
  
__Reproducibility__: 
* GitHub: lasseignelab/230313_TS_CCCinHumanAD
* Docker: tsoelter/rstudio_ccc_ad
    * Version: 1.0.5
* HPC: Yes
  * Resources: long partition (150hrs), 6 CPUs, 65GB per CPU

__Data__:
*Seurat objects*
* Name: geo_processed_seurat.rds, gse_processed_seurat.rds
* Location: data/seurat_preprocessing/
 
__Analysis Plan__:
* Load necessary packages 
* Load data
  * Seurat objects
* Create Single Cell Experiment objects
* Prepare for DGE Analysis
  * Pseudo-bulk data by cell type across samples
* DGE Analysis using DESeq2 between AD (group of interest) and CTRL
  * Wrapper function
  * Saves outputs
  
# Load packages
```{r}
# Time tracking
ptm <- proc.time()

# Setting seed
set.seed(42)

# Setting locale
options(encoding = "UTF-8")

# Load packages
suppressPackageStartupMessages({
  library(SingleCellExperiment)
  library(DESeq2)
  library(tidyverse)
  library(Seurat)
  library(here)
  library(styler)
  library(lintr)
})

source(here("src", "functions_CCCinAD_tms.R"))
```

# Load data
```{r}
geo_processed_seurat <- readRDS(here(
  "data",
  "seurat_preprocessing",
  "geo_processed_seurat.rds"
))

gse_processed_seurat <- readRDS(here(
  "data",
  "seurat_preprocessing",
  "gse_processed_seurat.rds"
))

object_list <- tibble::lst(
  geo_processed_seurat,
  gse_processed_seurat
)
```

# SCE objects creation
* pulling counts and metadata from seurat objects to create single cell experiment objects
* create the following columns in metadata beforehand:
  * sample_id (factorized sample column)
  * group_id (condition information - re-leveled for CTRL)
  * cluster_id (cell types)
```{r}
sce_objects <- make_sce(object_list)
```

# Pseudo-bulk
* This is done for each cell type across samples and condition by dataset
```{r}
counts_ls_list <- pseudobulk(sce_objects)
```
output:
[1] "geo_sce"
[1] 8
[1] 19
[1] "gse_sce"
[1] 7
[1] 21

# Get cell-type-specific metadata
```{r}
metadata_ls_list <- cts_metadata(
  object_list = sce_objects,
  counts_list = counts_ls_list
)
```
output:
[1] "geo_sce"
[1] 19  2
[1] "gse_sce"
[1] 21  2

# DE Analysis
* Using DESeq2, determine DEGs for each cell type between conditions
  * CTRL is our base level, while AD is the condition of interest
* Statistics: Wald test
* By dataset:
  * Dataset 1: geo
  * Dataset 2: gse
* Wrapper will also save DEGs to specified directory

Dataset 1
```{r}
# get cell types
cluster_names <- levels(colData(sce_objects$geo_sce)$cluster_id)

# get pseudo-bulked counts and their metadata
counts_ls <- counts_ls_list[["geo_counts_ls"]]
metadata_ls <- metadata_ls_list[["geo_metadata_ls"]]

# DEA using DESeq2
lapply(cluster_names,
  deseq2_dea,
  group_oi = "AD",
  B = "CTRL",
  path = "results/intermediate_outputs/geo/",
  metadata_ls = metadata_ls,
  counts_ls = counts_ls
)
```

Dataset 2
```{r}
# get cell types
cluster_names <- levels(colData(sce_objects$gse_sce)$cluster_id)

# get pseudo-bulked counts and their metadata
counts_ls <- counts_ls_list[["gse_counts_ls"]]
metadata_ls <- metadata_ls_list[["gse_metadata_ls"]]

# DEA using DESeq2
lapply(cluster_names,
  deseq2_dea,
  group_oi = "AD",
  B = "CTRL",
  path = "results/intermediate_outputs/gse/",
  metadata_ls = metadata_ls,
  counts_ls = counts_ls
)
```

# Session info
```{r}
sessionInfo() # see output below
```
R version 4.1.3 (2022-03-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.6 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
[1] C

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] lintr_3.0.2                 styler_1.9.1               
 [3] DESeq2_1.34.0               ggpubr_0.6.0               
 [5] cowplot_1.1.1               reshape2_1.4.4             
 [7] circlize_0.4.15             ComplexHeatmap_2.10.0      
 [9] here_1.0.1                  data.table_1.14.8          
[11] Matrix.utils_0.9.7          Matrix_1.5-4               
[13] SingleCellExperiment_1.16.0 SummarizedExperiment_1.24.0
[15] Biobase_2.54.0              GenomicRanges_1.46.1       
[17] GenomeInfoDb_1.30.1         IRanges_2.28.0             
[19] S4Vectors_0.32.4            BiocGenerics_0.40.0        
[21] MatrixGenerics_1.6.0        matrixStats_1.0.0          
[23] pheatmap_1.0.12             lubridate_1.9.2            
[25] forcats_1.0.0               stringr_1.5.0              
[27] dplyr_1.1.3                 purrr_1.0.2                
[29] readr_2.1.4                 tidyr_1.3.0                
[31] tibble_3.2.1                ggplot2_3.4.2              
[33] tidyverse_2.0.0             SeuratObject_4.1.3         
[35] Seurat_4.3.0.9011           OmnipathR_3.9.8            
[37] decoupleR_2.7.1            

loaded via a namespace (and not attached):
  [1] rappdirs_0.3.3            scattermore_1.2           R.methodsS3_1.8.2        
  [4] bit64_4.0.5               knitr_1.44                irlba_2.3.5.1            
  [7] DelayedArray_0.20.0       R.utils_2.12.2            KEGGREST_1.34.0          
 [10] RCurl_1.98-1.12           doParallel_1.0.17         generics_0.1.3           
 [13] callr_3.7.3               RSQLite_2.3.1             RANN_2.6.1               
 [16] future_1.33.0             bit_4.0.5                 tzdb_0.4.0               
 [19] spatstat.data_3.0-1       xml2_1.3.5                httpuv_1.6.11            
 [22] xfun_0.40                 hms_1.1.3                 evaluate_0.21            
 [25] promises_1.2.0.1          fansi_1.0.4               progress_1.2.2           
 [28] readxl_1.4.3              geneplotter_1.72.0        igraph_1.5.1             
 [31] DBI_1.1.3                 htmlwidgets_1.6.2         spatstat.geom_3.2-2      
 [34] ellipsis_0.3.2            selectr_0.4-2             backports_1.4.1          
 [37] annotate_1.72.0           deldir_1.0-9              sparseMatrixStats_1.6.0  
 [40] vctrs_0.6.3               remotes_2.4.2             ROCR_1.0-11              
 [43] abind_1.4-5               cachem_1.0.8              withr_2.5.0              
 [46] grr_0.9.5                 progressr_0.13.0          checkmate_2.2.0          
 [49] vroom_1.6.3               sctransform_0.3.5.9006    prettyunits_1.2.0        
 [52] goftest_1.2-3             cyclocomp_1.1.0           cluster_2.1.2            
 [55] lazyeval_0.2.2            crayon_1.5.2              genefilter_1.76.0        
 [58] spatstat.explore_3.2-1    edgeR_3.36.0              pkgconfig_2.0.3          
 [61] labeling_0.4.2            nlme_3.1-155              xmlparsedata_1.0.5       
 [64] rlang_1.1.1               globals_0.16.2            lifecycle_1.0.3          
 [67] miniUI_0.1.1.1            rex_1.2.1                 cellranger_1.1.0         
 [70] rprojroot_2.0.3           polyclip_1.10-4           lmtest_0.9-40            
 [73] carData_3.0-5             Rhdf5lib_1.16.0           zoo_1.8-12               
 [76] processx_3.8.1            ggridges_0.5.4            GlobalOptions_0.1.2      
 [79] png_0.1-8                 viridisLite_0.4.2         rjson_0.2.21             
 [82] bitops_1.0-7              R.oo_1.25.0               KernSmooth_2.23-20       
 [85] rhdf5filters_1.6.0        Biostrings_2.62.0         blob_1.2.4               
 [88] DelayedMatrixStats_1.16.0 shape_1.4.6               parallelly_1.36.0        
 [91] spatstat.random_3.1-5     R.cache_0.16.0            rstatix_0.7.2            
 [94] ggsignif_0.6.4            beachmat_2.10.0           scales_1.2.1             
 [97] memoise_2.0.1             magrittr_2.0.3            plyr_1.8.8               
[100] ica_1.0-3                 zlibbioc_1.40.0           compiler_4.1.3           
[103] dqrng_0.3.0               RColorBrewer_1.1-3        clue_0.3-64              
[106] fitdistrplus_1.1-11       cli_3.6.1                 XVector_0.34.0           
[109] listenv_0.9.0             ps_1.7.5                  patchwork_1.1.2          
[112] pbapply_1.7-2             MASS_7.3-55               tidyselect_1.2.0         
[115] stringi_1.7.12            yaml_2.3.7                locfit_1.5-9.8           
[118] ggrepel_0.9.3             tools_4.1.3               timechange_0.2.0         
[121] future.apply_1.11.0       parallel_4.1.3            rstudioapi_0.14          
[124] foreach_1.5.2             gridExtra_2.3             farver_2.1.1             
[127] Rtsne_0.16                DropletUtils_1.14.2       digest_0.6.33            
[130] shiny_1.7.4.1             Rcpp_1.0.11               car_3.1-2                
[133] broom_1.0.5               scuttle_1.4.0             later_1.3.1              
[136] RcppAnnoy_0.0.21          httr_1.4.7                AnnotationDbi_1.56.2     
[139] colorspace_2.1-0          rvest_1.0.3               XML_3.99-0.14            
[142] tensor_1.5                reticulate_1.30           splines_4.1.3            
[145] uwot_0.1.16               spatstat.utils_3.0-3      sp_2.0-0                 
[148] plotly_4.10.2             xtable_1.8-4              jsonlite_1.8.7           
[151] R6_2.5.1                  pillar_1.9.0              htmltools_0.5.6          
[154] mime_0.12                 glue_1.6.2                fastmap_1.1.1            
[157] BiocParallel_1.28.3       codetools_0.2-18          utf8_1.2.3               
[160] lattice_0.20-45           spatstat.sparse_3.0-2     logger_0.2.2             
[163] curl_5.0.2                leiden_0.4.3              waldo_0.5.1              
[166] survival_3.3-1            limma_3.50.3              rmarkdown_2.25           
[169] desc_1.4.2                munsell_0.5.0             GetoptLong_1.0.5         
[172] rhdf5_2.38.1              GenomeInfoDbData_1.2.7    iterators_1.0.14         
[175] HDF5Array_1.22.1          gtable_0.3.3

# Time tracking
```{r}
fptm <- proc.time() - ptm

# convert to minutes
(fptm[3] / 60)

# Time elapsed: 10 minutes
```

# Reproducibility
```{r}
# styler
style_file(here(
  "src",
  "biological_activity",
  "01_pseudobulk_dea.Rmd"
))

# lintr
lint(filename = here(
  "src",
  "biological_activity",
  "01_pseudobulk_dea.Rmd"
))
```
