---
title: "README"
author:
output: github_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# README Template 
**[Replace this header with your project name]**
Add link to pre-print here

## Authors 

**Tabea M. Soelter, Timothy C. Howton, Amanda D. Clark, Vishal H. Oza, Brittany N. Lasseigne**

The University of Alabama at Birmingham, Heersink School of Medicine, Department of Cell, Developmental and Integrative Biology

## Project Overview 
We used publicly available snRNA-seq AD data (Morabito et al., 2021) generated from postmortem human PFC to study altered glia-neuron interactions and their downstream effects in AD. We inferred differential CCC interactions between astrocytes, microglia, oligodendrocytes, or OPCs (sender cell types) and inhibitory or excitatory neurons (receiver cell types). We also investigated whether CCC shows high similarity across cell types by calculating the Jaccard Similarity Index (JI) of ligands, receptors, and target genes across cell types. Since CCC inference methodologies are known to produce false positives, we validated our interactions using an independent human PFC AD snRNA-seq dataset (Lau et al., 2020). We further investigated the resulting high-confidence ligand-receptor pairs from both data sets, their predicted downstream target genes, and signaling modulators through transcription factor (TF) and canonical signaling pathway activity.   
![alt text](https://github.com/lasseignelab/230313_TS_CCCinHumanAD/blob/main/results/figures/overview_figure1.png)

## Datasets
The datasets used in this study can be found on GEO:   
1. [Morabito et al., 2021](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE174367)   
2. [Lau et al., 2020](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157827)

## Scripts 
Data Alignment (Cell Ranger):
```{r echo=FALSE}
fs::dir_tree("src/cellranger/")
```
Ambient RNA removal (SoupX):
```{r echo=FALSE}
fs::dir_tree("src/soupX/")
```
Pre-processing (Seurat):
```{r echo=FALSE}
fs::dir_tree("src/seurat_preprocessing/")
```
Cell-cell communication inference (MultiNicheNet) and JI calculations:
```{r echo=FALSE}
fs::dir_tree("src/ccc/")
```
Biological Activity (decoupleR):
```{r echo=FALSE}
fs::dir_tree("src/biological_activity/")
```
Manuscript figures:
```{r echo=FALSE}
fs::dir_tree("src/manuscript_figures/")
```

## Lasseigne Lab 

[What is Happening in the Lasseigne Lab?](https://www.lasseigne.org/)

<img src="https://www.lasseigne.org/img/main/lablogo.png" width="75" height="75">

## Funding 

This work was supported in part by the UAB Lasseigne Lab funds, the NIA R00HG009678-04S1, and TMS through the Alzheimers of Central Alabama Lindy Harrell Predoctoral Scholar Program. 

## Acknowledgements 

We would also like to thank the members of the Lasseigne Lab, specifically Jordan H. Whitlock, Emma F. Jones, and Elizabeth J. Wilk for their valuable input.

## License 

This repository is licensed under the MIT License, see LICENSE documentation within this repository for more details.
[![MIT License](https://img.shields.io/badge/License-MIT-green.svg)](https://choosealicense.com/licenses/mit/)
