---
title: "01_geo_seurat_preprocessing"
author: "Tabea M. Soelter"
date: '2023-03-22'
output: html_document
---
**Pre-processing of snRNA-seq data from GEO (Accession #: GSE174367)**

__Goal__: Process GSE174367 (snRNA-seq) data for downstream analyses.
  
__Reproducibility__: 
* GitHub: lasseignelab/230313_TS_CCCinHumanAD
* Docker: tsoelter/rstudio_ccc_ad
    * Version: 1.0.1 
* HPC: Yes
  * Resources: long partition (150hrs), 6 CPUs, 65GB per CPU

__Data__: 
* Name: N/A
* Location: /data/user/tsoelter/projects/230313_TS_CCCinHumanAD/data/CellRangerCounts/GSE174367/
 
__Analysis Plan__: 
* Load necessary packages 
* Load data 
* Create Seurat object
* Perform quality control and filtering
* Integration
* Clustering and cell type assignment
* Save processed Seurat Object

__Analysis__:
# Load packages
```{r}
# Time tracking
ptm <- proc.time()

# Setting seed
set.seed(42)

# Load packages
suppressPackageStartupMessages({
  library(Seurat)
  library(tidyverse)
  library(harmony)
  library(clustree)
  library(stringr)
  library(scCustomize)
  library(here)
  library(styler)
  library(lintr)
})

source(here("src", "functions_CCCinAD_tms.R"))
```

# Create seurat object
```{r}
merged_seurat <- make_seurat_object(here(
  "data",
  "CellRangerCounts",
  "GSE174367"
))
```

# Calculate QC metrics
```{r}
merged_seurat <- calculate_qc(merged_seurat)
```

# Format metadata
```{r}
metadata <- format_metadata(merged_seurat)

# add sample information
metadata$sample <- ifelse(
  str_sub(merged_seurat@assays[["RNA"]]@data@Dimnames[[2]], 1, 4) == "CTRL",
  str_sub(merged_seurat@assays[["RNA"]]@data@Dimnames[[2]], 18, 24),
  str_sub(merged_seurat@assays[["RNA"]]@data@Dimnames[[2]], 16, 20)
)

metadata$sample <- gsub("_", "", metadata$sample)

dim(metadata) # 137754 7

# Add metadata to merged_seurat
merged_seurat@meta.data <- metadata
```

# Plot QC metrics
```{r warning=FALSE}
pdf(
  file = here(
    "results",
    "intermediate_outputs",
    "geo",
    "qc_plots.pdf"
  )
)
plot_qc(metadata)
dev.off()
```

# Filtering 
* Performing cell-level and gene-level filtering on the data. 
* Cell-level filtering is performed based on the plotted QC metrics (above).
* Gene-level filtering aims at removing count values = 0, since they can skew average gene expression
measurements.
  * Discussed in the following review: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02601-5)
```{r}
# Cell-level filtering
sub_seurat <- subset(merged_seurat,
  subset =
    nGene > 1000 &
      nGene < 10000 &
      mitoRatio < 0.20 &
      log10GenesPerUMI > 0.80
)

# Gene-level filtering
counts <- GetAssayData(sub_seurat, slot = "counts")

nonzero <- counts > 0

keep_genes <- rowSums(nonzero) >= 10

filtered_counts <- counts[keep_genes, ]

filtered_seurat <- CreateSeuratObject(filtered_counts,
  meta.data = sub_seurat@meta.data
)
```

# save filtered seurat object
```{r}
saveRDS(filtered_seurat, file = here(
  "data",
  "seurat_preprocessing",
  "geo_filtered_seurat.rds"
))
```

# Plot QC metrics after filtering
```{r warning=FALSE}
metadata <- filtered_seurat@meta.data
dim(metadata) # 65546 10

# plot and save post-filter QC outputs
pdf(file = here(
  "results",
  "intermediate_outputs",
  "geo",
  "qc_plots_filtered.pdf"
))
plot_qc(metadata)
dev.off()
```

# Determine cell cycle effects
```{r warning=FALSE}
pdf(
  file = here(
    "results",
    "intermediate_outputs",
    "geo",
    "cellcycle_pca.pdf"
  )
)
filtered_seurat <- cell_cylce_effects(filtered_seurat)
dev.off()
```

# Normalization and Integration
* We are using harmony as it has been shown to remove technical variation while preserving biological variation.
* We will have to provide the PCs from the elbow plot above
```{r}
filtered_seurat <- harmony_integration(filtered_seurat, dims = 1:20)
# harmony converged after 5 iterations
```

# Save integrated seurat object
```{r}
saveRDS(filtered_seurat, file = here(
  "data",
  "seurat_preprocessing",
  "geo_integrated_seurat.rds"
))
```

# Clustering
* Since I am using harmony for integration, I will have to ensure to use the RNA assay and the harmony reduction for everything downstream.
* I am using Leiden for clustering in my find_clusters function. 
```{r warning=FALSE}
# check active assay (needs to be RNA)
filtered_seurat@active.assay

# vector of desired resolutions
resolutions <- c(0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.2, 1.3)

# clustering of resolutions (Type "n" when asked about miniconda installation)
filtered_seurat <- find_clusters(filtered_seurat,
  dims = 1:20,
  reduction = "harmony",
  resolutions = resolutions
)
```

# Save clustered seurat object
```{r}
saveRDS(filtered_seurat, file = here(
  "data",
  "seurat_preprocessing",
  "geo_clustered_seurat.rds"
))
```

# Determine resolution
```{r}
# Plot resolutions
tree <- clustree(filtered_seurat@meta.data, prefix = "RNA_snn_res.")

pdf(
  file = here(
    "results",
    "intermediate_outputs",
    "geo",
    "clustree_umap.pdf"
  )
)

tree
# plotting UMAP for most stable resolution (1.2) for the number of cells
DimPlot(filtered_seurat,
  reduction = "umap_harmony",
  group.by = "RNA_snn_res.1.2"
)

dev.off()
```

# Find marker genes
```{r}
# Set the identity as Leiden with resolution 1.2
filtered_seurat <- SetIdent(filtered_seurat, value = "RNA_snn_res.1.2")

# Find marker genes
all_marker_genes <- FindAllMarkers(filtered_seurat,
  log2FC.threshold = 0.2,
  test.use = "wilcox",
  min.pct = 0.1,
  only.pos = TRUE,
  max.cells.per.ident = 50,
  assay = "RNA"
)

# Save marker genes
saveRDS(all_marker_genes, here(
  "results",
  "intermediate_outputs",
  "geo",
  "all_marker_genes.rds"
))

# Look at the top 25 up-regulated genes
top25 <- all_marker_genes %>%
  group_by(cluster) %>%
  top_n(-25, p_val_adj)

dim(top25) # 1251 7
```

# Feature plots to confirm clusters
```{r}
pdf(file = here("results", "intermediate_outputs", "geo", "feature_plots.pdf"))

FeaturePlot(filtered_seurat,
  features = c("PDGFRB"),
  label = TRUE
)

FeaturePlot(filtered_seurat,
  features = c("DOCK8", "APBB1IP", "ARHGAP15"),
  label = TRUE
)

FeaturePlot(filtered_seurat,
  features = c("SNAP25", "RBFOX1"),
  label = TRUE
)

FeaturePlot(filtered_seurat,
  features = c("AQP4", "GFAP", "ADGRV1"),
  label = TRUE
)

FeaturePlot(filtered_seurat,
  features = c("ST18", "MBP", "QDPR"),
  label = TRUE
)

FeaturePlot(filtered_seurat,
  features = c("PDGFRA", "PCDH15"),
  label = TRUE
)

FeaturePlot(filtered_seurat,
  features = c("FLT1", "VWF"),
  label = TRUE
)

dev.off()
```

# Assign cell types
```{r}
filtered_seurat <- RenameIdents(filtered_seurat,
  `1`  = "Oligodendrocytes",
  `2`  = "Oligodendrocytes",
  `3`  = "Oligodendrocytes",
  `4`  = "Oligodendrocytes",
  `5`  = "Microglia",
  `6`  = "OPCs",
  `7`  = "Astrocytes",
  `8`  = "Excitatory Neurons",
  `9`  = "Oligodendrocytes",
  `10` = "Astrocytes",
  `11` = "Inhibitory Neurons",
  `12` = "Inhibitory Neurons",
  `13` = "Inhibitory Neurons",
  `14` = "Excitatory Neurons",
  `15` = "Excitatory Neurons",
  `16` = "Oligodendrocytes",
  `17` = "Inhibitory Neurons",
  `18` = "Oligodendrocytes",
  `19` = "Excitatory Neurons",
  `20` = "Astrocytes",
  `21` = "Inhibitory Neurons",
  `22` = "Microglia",
  `23` = "Excitatory Neurons",
  `24` = "Excitatory Neurons",
  `25` = "Inhibitory Neurons",
  `26` = "Pericytes",
  `27` = "Excitatory Neurons",
  `28` = "Excitatory Neurons",
  `29` = "Endothelial cells",
  `30` = "Astrocytes",
  `31` = "Oligodendrocytes",
  `32` = "OPCs"
)

# UMAP of assigned clusters
png(file = here("results", "final_outputs", "geo", "UMAP.png"))
DimPlot_scCustom(filtered_seurat,
  figure_plot = TRUE,
  label = FALSE,
  ggplot_default_colors = TRUE
)
dev.off()

png(file = here("results", "intermediate_outputs", "geo", "UMAP_final.png"))
DimPlot(filtered_seurat)
dev.off()
```

# Save processed seurat object
```{r}
saveRDS(filtered_seurat, here(
  "data",
  "seurat_preprocessing",
  "geo_processed_seurat.rds"
))
```

# Session Info
```{r}
sessionInfo() # see output below
```
R version 4.1.3 (2022-03-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.6 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lintr_3.0.2        styler_1.9.1       here_1.0.1         scCustomize_1.1.1 
 [5] clustree_0.5.0     ggraph_2.1.0       harmony_0.1.0      Rcpp_1.0.10       
 [9] lubridate_1.9.2    forcats_1.0.0      stringr_1.5.0      dplyr_1.1.2       
[13] purrr_1.0.1        readr_2.1.4        tidyr_1.3.0        tibble_3.2.1      
[17] ggplot2_3.4.2      tidyverse_2.0.0    SeuratObject_4.1.3 Seurat_4.3.0.9002 

loaded via a namespace (and not attached):
  [1] backports_1.4.1        circlize_0.4.15        plyr_1.8.8            
  [4] igraph_1.4.2           lazyeval_0.2.2         sp_1.6-0              
  [7] splines_4.1.3          listenv_0.9.0          scattermore_0.8       
 [10] digest_0.6.31          htmltools_0.5.5        viridis_0.6.2         
 [13] fansi_1.0.4            checkmate_2.1.0        magrittr_2.0.3        
 [16] paletteer_1.5.0        tensor_1.5             cluster_2.1.2         
 [19] ROCR_1.0-11            limma_3.50.3           remotes_2.4.2         
 [22] tzdb_0.3.0             globals_0.16.2         graphlayouts_0.8.4    
 [25] matrixStats_0.63.0     R.utils_2.12.2         timechange_0.2.0      
 [28] spatstat.sparse_3.0-1  colorspace_2.1-0       rappdirs_0.3.3        
 [31] ggrepel_0.9.3          xfun_0.39              crayon_1.5.2          
 [34] callr_3.7.3            jsonlite_1.8.4         progressr_0.13.0      
 [37] spatstat.data_3.0-1    survival_3.3-1         zoo_1.8-12            
 [40] glue_1.6.2             polyclip_1.10-4        gtable_0.3.3          
 [43] leiden_0.4.3           R.cache_0.16.0         future.apply_1.10.0   
 [46] shape_1.4.6            abind_1.4-5            scales_1.2.1          
 [49] spatstat.random_3.1-4  miniUI_0.1.1.1         viridisLite_0.4.1     
 [52] xtable_1.8-4           reticulate_1.28        rex_1.2.1             
 [55] htmlwidgets_1.6.2      httr_1.4.5             RColorBrewer_1.1-3    
 [58] ellipsis_0.3.2         ica_1.0-3              R.methodsS3_1.8.2     
 [61] pkgconfig_2.0.3        farver_2.1.1           uwot_0.1.14           
 [64] deldir_1.0-6           utf8_1.2.3             janitor_2.2.0         
 [67] labeling_0.4.2         tidyselect_1.2.0       rlang_1.1.0           
 [70] reshape2_1.4.4         later_1.3.0            munsell_0.5.0         
 [73] tools_4.1.3            cli_3.6.1              ggprism_1.0.4         
 [76] generics_0.1.3         ggridges_0.5.4         evaluate_0.20         
 [79] fastmap_1.1.1          yaml_2.3.7             goftest_1.2-3         
 [82] rematch2_2.1.2         processx_3.8.1         knitr_1.42            
 [85] fitdistrplus_1.1-11    tidygraph_1.2.3        RANN_2.6.1            
 [88] pbapply_1.7-0          future_1.32.0          nlme_3.1-155          
 [91] mime_0.12              R.oo_1.25.0            ggrastr_1.0.1         
 [94] xml2_1.3.3             compiler_4.1.3         rstudioapi_0.14       
 [97] beeswarm_0.4.0         plotly_4.10.1          png_0.1-8             
[100] spatstat.utils_3.0-2   tweenr_2.0.2           stringi_1.7.12        
[103] ps_1.7.5               cyclocomp_1.1.0        desc_1.4.2            
[106] lattice_0.20-45        Matrix_1.5-4           vctrs_0.6.2           
[109] pillar_1.9.0           lifecycle_1.0.3        spatstat.geom_3.1-0   
[112] lmtest_0.9-40          GlobalOptions_0.1.2    RcppAnnoy_0.0.20      
[115] data.table_1.14.8      cowplot_1.1.1          irlba_2.3.5.1         
[118] httpuv_1.6.9           patchwork_1.1.2        R6_2.5.1              
[121] promises_1.2.0.1       KernSmooth_2.23-20     gridExtra_2.3         
[124] vipor_0.4.5            parallelly_1.35.0      codetools_0.2-18      
[127] MASS_7.3-55            rprojroot_2.0.3        withr_2.5.0           
[130] sctransform_0.3.5.9002 parallel_4.1.3         hms_1.1.3             
[133] grid_4.1.3             rmarkdown_2.21         snakecase_0.11.0      
[136] Rtsne_0.16             spatstat.explore_3.1-0 ggforce_0.4.1         
[139] shiny_1.7.4            ggbeeswarm_0.7.1

# Time tracking
```{r}
fptm <- proc.time() - ptm

# convert to hours
(fptm[3] / 60) / 60

# Time elapsed: 11.6 hours
# Hands-on time however is between 30-60 minutes!
```

# Reproducibility
```{r eval=FALSE, include=FALSE}
# styler
style_file(here(
  "src",
  "seurat_preprocessing",
  "01_geo_seurat_preprocessing.Rmd"
))

# lintr
lint(filename = here(
  "src",
  "seurat_preprocessing",
  "01_geo_seurat_preprocessing.Rmd"
))
```
