---
title: "01_vine_seurat_preprocessing"
author: "Tabea Soelter"
date: '2023-03-22'
output: html_document
---
**Pre-processing of VINE-seq snRNA-seq data from GEO (Accession #: GSE163577)**

__Goal__: Process GSE163577 (snRNA-seq) data for downstream analyses.
  
__Reproducibility__: 
* GitHub: lasseignelab/230313_TS_CCCinHumanAD
* Docker: tsoelter/rstudio_ccc_ad
    * Version: 1.0.0 
* HPC: Yes
  * Resources: long partition (150hrs), 6 CPUs, 65GB per CPU

__Data__: 
* Name: N/A
* Location: /data/user/tsoelter/projects/230313_TS_CCCinHumanAD/data/CellRangerCounts/VINE/
 
__Analysis Plan__: 
* Load necessary packages 
* Load data 
* Create seurat object
* Perform quality control and filtering
* Integration
* Clustering and cell type assignment
* Save processed seurat object

__Analysis__:
# Load packages
```{r}
set.seed(42)

library(Seurat)
library(tidyverse)
library(harmony)
library(clustree)
library(stringr)
library(scCustomize)
library(styler)
library(lintr)
library(here)
source(here("src", "functions_CCCinAD_tms.R"))
```

# Create seurat object
```{r}
merged_seurat <- make_seurat_object(here("data", "CellRangerCounts", "VINE"))
```

# Calculate QC metrics
```{r}
merged_seurat <- calculate_qc(merged_seurat)
```

# Format metadata
```{r}
metadata <- format_metadata(merged_seurat)
dim(metadata) # 402829 7
  
# Add metadata to merged_seurat
merged_seurat@meta.data <- metadata
```

# Plot QC metrics
```{r}
pdf(file = here("results", "intermediate_outputs", "vine", "qc_plots.pdf"))
plot_qc(metadata)
dev.off()
```

# Filtering 
* Performing cell-level and gene-level filtering on the data. 
* Cell-level filtering is performed based on the plotted QC metrics (above).
* Gene-level filtering aims at removing count values = 0, since they can skew average gene expression
measurements.
  * Discussed in the following review: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02601-5)
```{r}
# Cell-level filtering
sub_seurat <- subset(merged_seurat,
                     subset = 
                       nGene > 250 & 
                       nGene < 3000 & 
                       mitoRatio < 0.20 & 
                       log10GenesPerUMI > 0.80
                     )

# Gene-level filtering 
counts <- GetAssayData(sub_seurat, slot = "counts")

nonzero <- counts > 0

keep_genes <- rowSums(nonzero) >= 10

filtered_counts <- counts[keep_genes, ]

filtered_seurat <- CreateSeuratObject(filtered_counts, 
                                      meta.data = sub_seurat@meta.data)
```

# Save filtered seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "vine_filtered_seurat.rds"))
```

# Plot QC metrics after filtering
```{r}
metadata <- filtered_seurat@meta.data
dim(metadata) # 372166 10

# plot and save post-filter QC outputs
pdf(file = here("results", 
                "intermediate_outputs", 
                "vine", 
                "qc_plots_filtered.pdf"))
plot_qc(metadata)
dev.off()
```

# Determine cell cycle effects
```{r}
pdf(file = here("results", "intermediate_outputs", "vine", "cellcycle_pca.pdf"))
filtered_seurat <- cell_cylce_effects(filtered_seurat)
dev.off()
```

# Normalization and integration
* We are using harmony as it has been shown to remove technical variation while preserving biological variation.
* We will have to provide the PCs from the elbow plot above (1:20)
```{r}
filtered_seurat <- harmony_integration(filtered_seurat, dims = 1:20) # harmony converged after 7 iterations
```

# Save integrated seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "vine_integrated_seurat.rds"))
```

# Clustering
* Since I am using harmony for integration, I will have to ensure to use the RNA assay and the harmony reduction for everything downstream.
* I am using Leiden for clustering in my find_clusters function. 
```{r}
# check active assay (needs to be RNA)
filtered_seurat@active.assay 

# vector of desired resolutions
resolutions <- c(0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0)

# clustering of resolutions (when prompted about miniconda installation, answer with no)
filtered_seurat <- find_clusters(filtered_seurat, 
                                 dims = 1:20, 
                                 reduction = "harmony", 
                                 resolutions = resolutions)
```

# Save clustered seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "vine_clustered_seurat.rds"))
```

# Determine resolution
```{r}
# Plot resolutions between 0.1 and 0.9
tree <- clustree(filtered_seurat@meta.data, prefix = "RNA_snn_res.0.")

pdf(file = here("results", "intermediate_outputs", "vine", "clustree_umap.pdf"))
tree

# plotting UMAP for most stable resolution (0.2)
DimPlot(filtered_seurat, 
        reduction = "umap_harmony", 
        group.by = "RNA_snn_res.0.2")

dev.off()
```

# Find marker genes
```{r}
# Set the identity as Leiden with resolution 0.2
filtered_seurat <- SetIdent(filtered_seurat, value = "RNA_snn_res.0.2")

# Find marker genes
all_marker_genes <- FindAllMarkers(filtered_seurat, 
                                   log2FC.threshold = 0.2, 
                                   test.use = "wilcox",
                                   min.pct = 0.1,
                                   only.pos = TRUE,
                                   max.cells.per.ident = 50,
                                   assay = "RNA")

# Save marker genes
saveRDS(all_marker_genes, here("results", 
                               "intermediate_outputs", 
                               "vine", 
                               "all_marker_genes.rds")
        )

# Look at the top 25 up-regulated genes
top25 <- all_marker_genes %>%
  group_by(cluster) %>%
  top_n(-25, p_val_adj)

dim(top25) # 428 7
```

# Finding markers of unresolved clusters - EDIT THIS
* Clusters unidentified are: 18-21
```{r}
# vector of clusters of interest
identities <- c("18", "19", "20", "21")

# find marker genes across clusters of interest
top_markers <- find_markers(filtered_seurat, 
                            resolution = "RNA_snn_res.0.2", 
                            identities = identities, 
                            value = 15)
```

# Assign cell types
```{r}
filtered_seurat <- RenameIdents(filtered_seurat, 
                                `1`  = "Oligodendrocytes",
                                `2`  = "Oligodendrocytes",
                                `3`  = "Astrocytes",
                                `4`  = "Microglia",
                                `5`  = "Excitatory Neurons",
                                `6`  = "OPCs",
                                `7`  = "Inhibitory Neurons",
                                `8`  = "Inhibitory Neurons",
                                `9`  = "Oligodendrocytes",
                                `10` = "Inhibitory Neurons", 
                                `11` = "Astrocytes", 
                                `12` = "Excitatory Neurons",
                                `13` = "Excitatory Neurons",
                                `14` = "Excitatory Neurons",
                                `15` = "Inhibitory Neurons",
                                `16` = "Inhibitory Neurons",
                                `17` = "Excitatory Neurons",
                                `18` = "Microglia",
                                `19` = "Oligodendrocytes",
                                `20` = "Microglia",
                                `21` = "Microglia")
```

# Save processed seurat object
```{r}
saveRDS(filtered_seurat, here("data", "vine_processed_seurat.rds"))
```

# Session info
```{r}
sessionInfo() # see output below
```
R version 4.1.3 (2022-03-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.5 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] here_1.0.1         lintr_3.0.2        styler_1.9.1       scCustomize_1.1.1  clustree_0.5.0    
 [6] ggraph_2.1.0       harmony_0.1.0      Rcpp_1.0.10        lubridate_1.9.2    forcats_1.0.0     
[11] stringr_1.5.0      dplyr_1.1.0        purrr_1.0.1        readr_2.1.4        tidyr_1.3.0       
[16] tibble_3.2.0       ggplot2_3.4.1      tidyverse_2.0.0    SeuratObject_4.1.3 Seurat_4.3.0.9002 

loaded via a namespace (and not attached):
  [1] backports_1.4.1        circlize_0.4.15        plyr_1.8.8             igraph_1.4.1          
  [5] lazyeval_0.2.2         sp_1.6-0               splines_4.1.3          listenv_0.9.0         
  [9] scattermore_0.8        digest_0.6.31          htmltools_0.5.4        viridis_0.6.2         
 [13] fansi_1.0.4            checkmate_2.1.0        magrittr_2.0.3         paletteer_1.5.0       
 [17] tensor_1.5             cluster_2.1.2          ROCR_1.0-11            limma_3.50.3          
 [21] remotes_2.4.2          tzdb_0.3.0             globals_0.16.2         graphlayouts_0.8.4    
 [25] matrixStats_0.63.0     R.utils_2.12.2         timechange_0.2.0       spatstat.sparse_3.0-1 
 [29] colorspace_2.1-0       rappdirs_0.3.3         ggrepel_0.9.3          xfun_0.37             
 [33] crayon_1.5.2           callr_3.7.3            jsonlite_1.8.4         progressr_0.13.0      
 [37] spatstat.data_3.0-1    survival_3.3-1         zoo_1.8-11             glue_1.6.2            
 [41] polyclip_1.10-4        gtable_0.3.1           leiden_0.4.3           R.cache_0.16.0        
 [45] future.apply_1.10.0    shape_1.4.6            abind_1.4-5            scales_1.2.1          
 [49] spatstat.random_3.1-3  miniUI_0.1.1.1         viridisLite_0.4.1      xtable_1.8-4          
 [53] reticulate_1.28        rex_1.2.1              htmlwidgets_1.6.1      httr_1.4.5            
 [57] RColorBrewer_1.1-3     ellipsis_0.3.2         ica_1.0-3              R.methodsS3_1.8.2     
 [61] pkgconfig_2.0.3        farver_2.1.1           uwot_0.1.14            deldir_1.0-6          
 [65] utf8_1.2.3             janitor_2.2.0          labeling_0.4.2         tidyselect_1.2.0      
 [69] rlang_1.0.6            reshape2_1.4.4         later_1.3.0            munsell_0.5.0         
 [73] tools_4.1.3            cli_3.6.0              ggprism_1.0.4          generics_0.1.3        
 [77] ggridges_0.5.4         evaluate_0.20          fastmap_1.1.1          yaml_2.3.7            
 [81] goftest_1.2-3          rematch2_2.1.2         processx_3.8.0         knitr_1.42            
 [85] fitdistrplus_1.1-8     tidygraph_1.2.3        RANN_2.6.1             pbapply_1.7-0         
 [89] future_1.32.0          nlme_3.1-155           mime_0.12              R.oo_1.25.0           
 [93] ggrastr_1.0.1          xml2_1.3.3             compiler_4.1.3         rstudioapi_0.14       
 [97] beeswarm_0.4.0         plotly_4.10.1          png_0.1-8              spatstat.utils_3.0-2  
[101] tweenr_2.0.2           stringi_1.7.12         ps_1.7.2               cyclocomp_1.1.0       
[105] desc_1.4.2             lattice_0.20-45        Matrix_1.5-3           vctrs_0.5.2           
[109] pillar_1.8.1           lifecycle_1.0.3        spatstat.geom_3.1-0    lmtest_0.9-40         
[113] GlobalOptions_0.1.2    RcppAnnoy_0.0.20       data.table_1.14.8      cowplot_1.1.1         
[117] irlba_2.3.5.1          httpuv_1.6.9           patchwork_1.1.2        R6_2.5.1              
[121] promises_1.2.0.1       KernSmooth_2.23-20     gridExtra_2.3          vipor_0.4.5           
[125] parallelly_1.34.0      codetools_0.2-18       MASS_7.3-55            rprojroot_2.0.3       
[129] withr_2.5.0            sctransform_0.3.5.9002 mgcv_1.8-39            parallel_4.1.3        
[133] hms_1.1.2              grid_4.1.3             rmarkdown_2.20         snakecase_0.11.0      
[137] Rtsne_0.16             spatstat.explore_3.0-6 ggforce_0.4.1          shiny_1.7.4           
[141] ggbeeswarm_0.7.1      

# Reproducibility
```{r}
# styler
style_file(here("src",
                "seurat_preprocessing",
                "01_vine_seurat_preprocessing.Rmd"))

# lintr
lint(filename = here("src",
                     "seurat_preprocessing",
                     "01_vine_seurat_preprocessing.Rmd"))
```