---
title: "01_ambientRNA_removal"
author: "Tabea M. Soelter"
date: '2023-07-17'
output: html_document
---
**Ambient RNA removal in snRNA-seq data from GEO (Accession #: GSE157827)**

__Goal__: Remove ambient RNA from GSE157827 (snRNA-seq) data for downstream analyses, especially since 4 samples returned the ambient RNA warning after alignment with Cell Ranger. 
  
__Reproducibility__: 
* GitHub: lasseignelab/230313_TS_CCCinHumanAD
* Docker: tsoelter/rstudio_ccc_ad
    * Version: 1.0.3 
* HPC: Yes
  * Resources: long partition (150hrs), 8 CPUs, 65GB per CPU

__Data__: 
* Name: N/A
* Location: /data/user/tsoelter/projects/230313_TS_CCCinHumanAD/data/CellRangerCounts/GSE157827/pre_soupX/
 
__Analysis Plan__: 
* Load necessary packages 
* Load h5 data from Cell Ranger
* Make Soup Channel object
* Create Seurat object
* Cluster Seurat object
* Profile the soup
* Create integer matrix
* Save filtered objects

__Analysis__:
# load packages
```{r}
# Time tracking
ptm <- proc.time()

# Setting seed
set.seed(42)

# Load packages
suppressPackageStartupMessages({
  library(Seurat)
  library(SoupX)
  library(DropletUtils)
  library(here)
  library(styler)
  library(lintr)
})

source(here("src", "functions_CCCinAD_tms.R"))
```

# Ambient RNA removal
* Here I am using a function which takes in the h5 files from Cell Ranger and performs all necessary steps to remove ambient RNA. Ultimately, filtered files necessary for Seurat preprocessing are saved by sample.
```{r}
remove_ambientRNA(
  inputs = here(
    "data",
    "CellRangerCounts",
    "GSE157827",
    "pre_soupX/"
  ),
  outputs = here(
    "data",
    "CellRangerCounts",
    "GSE157827",
    "post_soupX/"
  ),
  plots = here(
    "results",
    "intermediate_outputs",
    "gse",
    "soupX/"
  )
)
```

# Session Info
```{r}
sessionInfo() # see output below
```
R version 4.1.3 (2022-03-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.6 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
[1] C

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lintr_3.0.2                 styler_1.9.1                here_1.0.1                 
 [4] DropletUtils_1.14.2         SingleCellExperiment_1.16.0 SummarizedExperiment_1.24.0
 [7] Biobase_2.54.0              GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        
[10] IRanges_2.28.0              S4Vectors_0.32.4            BiocGenerics_0.40.0        
[13] MatrixGenerics_1.6.0        matrixStats_1.0.0           SoupX_1.6.2                
[16] SeuratObject_4.1.3          Seurat_4.3.0.9011          

loaded via a namespace (and not attached):
  [1] plyr_1.8.8                igraph_1.5.0              lazyeval_0.2.2           
  [4] sp_2.0-0                  splines_4.1.3             BiocParallel_1.28.3      
  [7] listenv_0.9.0             scattermore_1.2           ggplot2_3.4.2            
 [10] digest_0.6.33             htmltools_0.5.5           fansi_1.0.4              
 [13] magrittr_2.0.3            tensor_1.5                cluster_2.1.2            
 [16] ROCR_1.0-11               remotes_2.4.2             limma_3.50.3             
 [19] globals_0.16.2            R.utils_2.12.2            spatstat.sparse_3.0-2    
 [22] colorspace_2.1-0          ggrepel_0.9.3             xfun_0.39                
 [25] dplyr_1.1.2               crayon_1.5.2              callr_3.7.3              
 [28] RCurl_1.98-1.12           jsonlite_1.8.7            progressr_0.13.0         
 [31] spatstat.data_3.0-1       survival_3.3-1            zoo_1.8-12               
 [34] glue_1.6.2                polyclip_1.10-4           gtable_0.3.3             
 [37] zlibbioc_1.40.0           XVector_0.34.0            leiden_0.4.3             
 [40] DelayedArray_0.20.0       R.cache_0.16.0            Rhdf5lib_1.16.0          
 [43] future.apply_1.11.0       HDF5Array_1.22.1          abind_1.4-5              
 [46] scales_1.2.1              edgeR_3.36.0              spatstat.random_3.1-5    
 [49] miniUI_0.1.1.1            Rcpp_1.0.11               viridisLite_0.4.2        
 [52] xtable_1.8-4              reticulate_1.30           dqrng_0.3.0              
 [55] bit_4.0.5                 rex_1.2.1                 htmlwidgets_1.6.2        
 [58] httr_1.4.6                RColorBrewer_1.1-3        ellipsis_0.3.2           
 [61] ica_1.0-3                 pkgconfig_2.0.3           R.methodsS3_1.8.2        
 [64] scuttle_1.4.0             uwot_0.1.16               deldir_1.0-9             
 [67] locfit_1.5-9.8            utf8_1.2.3                tidyselect_1.2.0         
 [70] rlang_1.1.1               reshape2_1.4.4            later_1.3.1              
 [73] munsell_0.5.0             tools_4.1.3               cli_3.6.1                
 [76] generics_0.1.3            ggridges_0.5.4            stringr_1.5.0            
 [79] fastmap_1.1.1             goftest_1.2-3             bit64_4.0.5              
 [82] processx_3.8.1            knitr_1.43                fitdistrplus_1.1-11      
 [85] purrr_1.0.1               RANN_2.6.1                pbapply_1.7-2            
 [88] future_1.33.0             nlme_3.1-155              sparseMatrixStats_1.6.0  
 [91] mime_0.12                 R.oo_1.25.0               xml2_1.3.3               
 [94] hdf5r_1.3.8               compiler_4.1.3            rstudioapi_0.14          
 [97] plotly_4.10.2             png_0.1-8                 spatstat.utils_3.0-3     
[100] tibble_3.2.1              stringi_1.7.12            ps_1.7.5                 
[103] cyclocomp_1.1.0           desc_1.4.2                lattice_0.20-45          
[106] Matrix_1.5-4              vctrs_0.6.3               pillar_1.9.0             
[109] lifecycle_1.0.3           rhdf5filters_1.6.0        spatstat.geom_3.2-2      
[112] lmtest_0.9-40             RcppAnnoy_0.0.21          data.table_1.14.8        
[115] cowplot_1.1.1             bitops_1.0-7              irlba_2.3.5.1            
[118] httpuv_1.6.11             patchwork_1.1.2           R6_2.5.1                 
[121] promises_1.2.0.1          KernSmooth_2.23-20        gridExtra_2.3            
[124] parallelly_1.36.0         codetools_0.2-18          MASS_7.3-55              
[127] rhdf5_2.38.1              rprojroot_2.0.3           withr_2.5.0              
[130] sctransform_0.3.5.9006    GenomeInfoDbData_1.2.7    parallel_4.1.3           
[133] grid_4.1.3                beachmat_2.10.0           tidyr_1.3.0              
[136] DelayedMatrixStats_1.16.0 Rtsne_0.16                spatstat.explore_3.2-1   
[139] shiny_1.7.4.1

# Time tracking
```{r}
fptm <- proc.time() - ptm

# convert to minutes
(fptm[3] / 60)

# Time elapsed: 1 hour
# Hands-on time however is 5 minutes!
```

# Reproducibility
```{r eval=FALSE, include=FALSE}
# styler
style_file(here(
  "src",
  "soupX",
  "01_ambientRNA_removal.Rmd"
))

# lintr
lint(filename = here(
  "src",
  "soupX",
  "01_ambientRNA_removal.Rmd"
))
```

