---
title: "01_integration"
author: "Tabea M. Soelter"
date: '2023-05-30'
output: html_document
---
# Integration of 8v8 and cortex v cotex datasets
* Here is crude code to integrate both 8v8 (VINE-seq) and CvC (vine_cortex vs geo)
* I am processing all the way to cell type assignment and saving processed seurat objects.

__Analysis__:
# Load packages
```{r}
# Time tracking
ptm <- proc.time()

# Setting seed
set.seed(42)

# Load packages
suppressPackageStartupMessages({
  library(Seurat)
  library(tidyverse)
  library(harmony)
  library(clustree)
  library(stringr)
  library(scCustomize)
  library(here)
  library(styler)
  library(lintr)
})

source(here("src", "functions_CCCinAD_tms.R"))
```
# Editing make seurat object function
* I want to add brain region information to the metadata
```{r}
make_seurat_object2 <- function(path){
  counts_list <- list.dirs(path, 
                           full.names = TRUE, 
                           recursive = FALSE)
  object_list <- vector('list')
  for (i in counts_list){
    counts <- Read10X(i)
    print(i)
    sample_name <- basename(i)
    sample_name <- gsub("[[:punct:]]", "", sample_name)
    seurat_object <- CreateSeuratObject(counts = counts, project = sample_name, min.features = 200)
    if (str_sub(sample_name, - 2, - 1) == "AD") {
      seurat_object$orig.ident <- "AD"
    } else {
      seurat_object$orig.ident <- "CTRL"
    }
    if (str_sub(sample_name, 15, 17) == "hip") {
      seurat_object$region <- "hippocampus"
    } else {
      seurat_object$region <- "cortex"
    }
    object <- seurat_object
    object_list[[i]] <- object
  } 
  print("Making diseased Seurat Object")
  AD_list <- object_list[grepl("AD", names(object_list))]
  for (i in names(AD_list)) {
    sample_name <- basename(i)
    sample_name <- gsub("[[:punct:]]", "", sample_name)
    AD_list[[i]] <- RenameCells(AD_list[[i]],
                                add.cell.id = sample_name)
  }
  diseased <- Merge_Seurat_List(AD_list)
  
  print("Making control Seurat Object")
  CTRL_list <- object_list[grepl("CTRL", names(object_list))]
  for (i in names(CTRL_list)) {
    sample_name <- basename(i)
    sample_name <- gsub("[[:punct:]]", "", sample_name)
    CTRL_list[[i]] <- RenameCells(CTRL_list[[i]],
                                  add.cell.id = sample_name)
  }
  control <- Merge_Seurat_List(CTRL_list)
  
  print("Making merged Seurat Object")
  merged_seurat <- merge(x = control,
                         y = diseased,
                         add.cell.id = c("CTRL", "AD"))
  return(merged_seurat)
}
```

# Create seurat object
```{r}
merged_seurat <- make_seurat_object2(here("data", "CellRangerCounts", "test", "8v8"))
```

# Calculate QC metrics
```{r}
merged_seurat <- calculate_qc(merged_seurat)
```

# Format metadata
```{r}
metadata <- format_metadata(merged_seurat)
dim(metadata) # 381851 8

# Add metadata to merged_seurat
merged_seurat@meta.data <- metadata
```

# Plot QC metrics
```{r warning=FALSE}
#pdf(file = here("results", "intermediate_outputs", "", "qc_plots.pdf"))
plot_qc(metadata)
#dev.off()
```

# Filtering 
* Performing cell-level and gene-level filtering on the data. 
* Cell-level filtering is performed based on the plotted QC metrics (above).
* Gene-level filtering aims at removing count values = 0, since they can skew average gene expression
measurements.
  * Discussed in the following review: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02601-5)
```{r}
# Cell-level filtering
sub_seurat <- subset(merged_seurat,
  subset =
    nGene > 300 &
      nGene < 3000 &
      mitoRatio < 0.20 &
      log10GenesPerUMI > 0.80
)

# Gene-level filtering
counts <- GetAssayData(sub_seurat, slot = "counts")

nonzero <- counts > 0

keep_genes <- rowSums(nonzero) >= 10

filtered_counts <- counts[keep_genes, ]

filtered_seurat <- CreateSeuratObject(filtered_counts,
  meta.data = sub_seurat@meta.data
)
```

# Save filtered seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "test", "vine_filtered_seurat.rds"))
```

# Plot QC metrics after filtering
```{r warning=FALSE}
metadata <- filtered_seurat@meta.data
dim(metadata) # 325553 11

# plot and save post-filter QC outputs
#pdf(file = here(
#  "results",
#  "intermediate_outputs",
#  "",
#  "qc_plots_filtered.pdf"
#))
plot_qc(metadata)
#dev.off()
```

# Determine cell cycle effects
```{r warning=FALSE}
#pdf(file = here("results", "intermediate_outputs", "", "cellcycle_pca.pdf"))
filtered_seurat <- cell_cylce_effects(filtered_seurat)
#dev.off()
```

# Normalization and integration
* We are using harmony as it has been shown to remove technical variation while preserving biological variation.
* We will have to provide the PCs from the elbow plot above (1:25)
```{r warning=FALSE}
filtered_seurat <- harmony_integration(filtered_seurat, dims = 1:25)
# harmony converged after 2 iterations
```

# Save integrated seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "test", "vine_integrated_seurat.rds"))
```

# Clustering
* Since I am using harmony for integration, I will have to ensure to use the RNA assay and the harmony reduction for everything downstream.
* I am using Leiden for clustering in my find_clusters function. 
```{r warning=FALSE}
# check active assay (needs to be RNA)
filtered_seurat@active.assay

# vector of desired resolutions
resolutions <- c(0.4, 0.6, 0.8, 1.0, 1.2, 1.3, 1.4, 1.5, 1.6)

# clustering of resolutions (answer no to miniconda installation)
filtered_seurat <- find_clusters(filtered_seurat,
  dims = 1:25,
  reduction = "harmony",
  resolutions = resolutions
)
```

# Save clustered seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "test", "vine_clustered_seurat.rds"))
```

# Determine resolution
```{r}
# Plot resolutions between 0.1 and 0.9
tree <- clustree(filtered_seurat@meta.data, prefix = "RNA_snn_res.0.")

# Resolutions between 1 and 1.9
tree2 <- clustree(filtered_seurat@meta.data, prefix = "RNA_snn_res.1.")

#pdf(file = here("results", "intermediate_outputs", "", "clustree_umap.pdf"))
tree
tree2

# plotting UMAP for most stable resolution (1.5)
DimPlot(filtered_seurat,
  reduction = "umap_harmony",
  group.by = "RNA_snn_res.1.5"
)

DimPlot(vine_clustered_seurat,
  reduction = "umap_harmony",
  group.by = "RNA_snn_res.1.5"
)

DimPlot(filtered_seurat,
  reduction = "umap_harmony",
  group.by = "region",
  shuffle = TRUE
)

DimPlot(filtered_seurat,
  reduction = "umap_harmony",
  group.by = "region",
  shuffle = FALSE
)

table(filtered_seurat@meta.data$region)
#dev.off()
```

# Find marker genes
```{r}
# Set the identity as Leiden with resolution 0.7
filtered_seurat <- SetIdent(filtered_seurat, value = "RNA_snn_res.1.5")

# Find marker genes
all_marker_genes <- FindAllMarkers(filtered_seurat,
  log2FC.threshold = 0.2,
  test.use = "wilcox",
  min.pct = 0.1,
  only.pos = TRUE,
  max.cells.per.ident = 50,
  assay = "RNA"
)

# Save marker genes
saveRDS(all_marker_genes, here(
  "data",
  "test",
  "vine_all_marker_genes.rds"
))

# Look at the top 25 up-regulated genes
top25 <- all_marker_genes %>%
  group_by(cluster) %>%
  top_n(-25, p_val_adj)

dim(top25) # 1837 7
```

# Feature Plots
```{r}
FeaturePlot(vine_clustered_seurat, features = c("PDGFRB"), label = TRUE)

FeaturePlot(vine_clustered_seurat, features = c("DOCK8", "APBB1IP", "ARHGAP15"), label = TRUE)

FeaturePlot(vine_clustered_seurat, features = c("SNAP25", "RBFOX1"), label = TRUE)

FeaturePlot(vine_clustered_seurat, features = c("SLC7A11", "GFAP"), label = TRUE)

FeaturePlot(vine_clustered_seurat, features = c("ST18", "MBP", "IL1RAPL1"), label = TRUE)

FeaturePlot(vine_clustered_seurat, features = c("TNR", "PCDH15", "OPCML"), label = TRUE)

FeaturePlot(vine_clustered_seurat, features = c("FLT1", "VWF"), label = TRUE)

FeaturePlot(vine_clustered_seurat, features = c("SSTR3"), label = FALSE)


# find marker genes across clusters of interest
#top_markers <- find_markers(vine_clustered_seurat,
#  resolution = "RNA_snn_res.1.5",
#  identities = "42",
#  value = 25
#)
```

# Assign cell types
```{r}
vine_clustered_seurat <- RenameIdents(vine_clustered_seurat,
  `1`  = "Oligodendrocytes",
  `2`  = "Endothelial cells",
  `3`  = "Pericytes",
  `4`  = "Oligodendrocytes",
  `5`  = "Astrocytes",
  `6`  = "Endothelial cells",
  `7`  = "Oligodendrocytes",
  `8`  = "Endothelial cells",
  `9`  = "Oligodendrocytes",
  `10` = "Pericytes",
  `11` = "Endothelial cells",
  `12` = "Pericytes",
  `13` = "Endothelial cells",
  `14` = "Pericytes",
  `15` = "Endothelial cells",
  `16` = "Pericytes",
  `17` = "Endothelial cells",
  `18` = "Pericytes",
  `19` = "Microglia",
  `20` = "Astrocytes",
  `21` = "Astrocytes",
  `22` = "Neurons",
  `23` = "Oligodendrocytes",
  `24` = "Neurons",
  `25` = "Endothelial cells",
  `26` = "Pericytes",
  `27` = "Oligodendrocytes",
  `28` = "Endothelial cells",
  `29` = "Endothelial cells",
  `30` = "Astrocytes",
  `31` = "OPCs",
  `32` = "Astrocytes",
  `33` = "Astrocytes",
  `34` = "Microglia",
  `35` = "Microglia",
  `36` = "Microglia",
  `37` = "Excitatory Neurons 2",
  `38` = "Neurons",
  `39` = "Microglia",
  `40` = "Astrocytes",
  `41` = "Astrocytes",
  `42` = "Inhibitory Neurons 2",
  `43` = "Epithelial cells",
  `44` = "Astrocytes"
)
```

# Subclustering of cluster 18
```{r}
# Subcluster the neuron clusters
vine_clustered_seurat <- FindSubCluster(vine_clustered_seurat,
  cluster = "Neurons",
  graph.name = "RNA_snn",
  resolution = 0.2
)

# Set identity to resolution with sub clusters
vine_clustered_seurat <- SetIdent(vine_clustered_seurat, value = "sub.cluster")

# Sub-clusters for marker identification
idents <- c("Neurons_0", "Neurons_1", "Neurons_2", "Neurons_3", "Neurons_4", "Neurons_5", "Neurons_6", "Neurons_7", "Neurons_8")

# Find markers for sub-clusters
neurons_markers <- find_markers(vine_clustered_seurat,
  resolution = "sub.cluster",
  identities = idents,
  value = 5
)

dim(neurons_markers) # 45 6

DimPlot(vine_clustered_seurat, label = TRUE)

# Assign new cell types
filtered_seurat_test <- RenameIdents(vine_clustered_seurat,
  "Neurons_0" = "Microglia",
  "Neurons_1" = "Microglia",
  "Neurons_2" = "Excitatory Neurons",
  "Neurons_3" = "Inhibitory Neurons",
  "Neurons_4" = "Inhibitory Neurons",
  "Neurons_5" = "Inhibitory Neurons",
  "Neurons_6" = "Microglia",
  "Neurons_7" = "Excitatory Neurons",
  "Neurons_8" = "Microglia"
)

DimPlot(filtered_seurat_test)
```
# save object
```{r}
saveRDS(filtered_seurat_test, file = here("data", "test", "vine_processed_seurat.rds"))
```

_______________________________________________________________________________

# Cortex
# Editing make seurat object function
* I want to add brain region information to the metadata
```{r}
make_seurat_object3 <- function(path){
  counts_list <- list.dirs(path, 
                           full.names = TRUE, 
                           recursive = FALSE)
  object_list <- vector('list')
  for (i in counts_list){
    counts <- Read10X(i)
    print(i)
    sample_name <- basename(i)
    sample_name <- gsub("[[:punct:]]", "", sample_name)
    seurat_object <- CreateSeuratObject(counts = counts, project = sample_name, min.features = 200)
    if (str_sub(sample_name, - 2, - 1) == "AD") {
      seurat_object$orig.ident <- "AD"
    } else {
      seurat_object$orig.ident <- "CTRL"
    }
    if (str_sub(sample_name, 16, 18) == "geo") {
      seurat_object$dataset <- "geo"
    } else {
      seurat_object$dataset <- "vine"
    }
    object <- seurat_object
    object_list[[i]] <- object
  } 
  print("Making diseased Seurat Object")
  AD_list <- object_list[grepl("AD", names(object_list))]
  for (i in names(AD_list)) {
    sample_name <- basename(i)
    sample_name <- gsub("[[:punct:]]", "", sample_name)
    AD_list[[i]] <- RenameCells(AD_list[[i]],
                                add.cell.id = sample_name)
  }
  diseased <- Merge_Seurat_List(AD_list)
  
  print("Making control Seurat Object")
  CTRL_list <- object_list[grepl("CTRL", names(object_list))]
  for (i in names(CTRL_list)) {
    sample_name <- basename(i)
    sample_name <- gsub("[[:punct:]]", "", sample_name)
    CTRL_list[[i]] <- RenameCells(CTRL_list[[i]],
                                  add.cell.id = sample_name)
  }
  control <- Merge_Seurat_List(CTRL_list)
  
  print("Making merged Seurat Object")
  merged_seurat <- merge(x = control,
                         y = diseased,
                         add.cell.id = c("CTRL", "AD"))
  return(merged_seurat)
}
```

# Create seurat object
```{r}
merged_seurat <- make_seurat_object3(here("data", "CellRangerCounts", "test", "cortex"))
```

# Calculate QC metrics
```{r}
merged_seurat <- calculate_qc(merged_seurat)
```

# Format metadata
```{r}
metadata <- format_metadata(merged_seurat)
dim(metadata) # 431114 8

# Add metadata to merged_seurat
merged_seurat@meta.data <- metadata
```

# Plot QC metrics
```{r warning=FALSE}
#pdf(file = here("results", "intermediate_outputs", "", "qc_plots.pdf"))
plot_qc(metadata)
#dev.off()
```

# Filtering 
* Performing cell-level and gene-level filtering on the data. 
* Cell-level filtering is performed based on the plotted QC metrics (above).
* Gene-level filtering aims at removing count values = 0, since they can skew average gene expression
measurements.
  * Discussed in the following review: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02601-5)
```{r}
# Cell-level filtering
sub_seurat <- subset(merged_seurat,
  subset =
    nGene > 300 &
      nGene < 10000 &
      mitoRatio < 0.20 &
      log10GenesPerUMI > 0.80
)

# Gene-level filtering
counts <- GetAssayData(sub_seurat, slot = "counts")

nonzero <- counts > 0

keep_genes <- rowSums(nonzero) >= 10

filtered_counts <- counts[keep_genes, ]

filtered_seurat <- CreateSeuratObject(filtered_counts,
  meta.data = sub_seurat@meta.data
)
```

# Save filtered seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "test", "cortex_filtered_seurat.rds"))
```

# Plot QC metrics after filtering
```{r warning=FALSE}
metadata <- filtered_seurat@meta.data
dim(metadata) # 343335 11

# plot and save post-filter QC outputs
#pdf(file = here(
#  "results",
#  "intermediate_outputs",
#  "",
#  "qc_plots_filtered.pdf"
#))
plot_qc(metadata)
#dev.off()
```

# Determine cell cycle effects
```{r warning=FALSE}
#pdf(file = here("results", "intermediate_outputs", "", "cellcycle_pca.pdf"))
filtered_seurat <- cell_cylce_effects(filtered_seurat)
#dev.off()
```

# Normalization and integration
* We are using harmony as it has been shown to remove technical variation while preserving biological variation.
* We will have to provide the PCs from the elbow plot above (1:20)
```{r warning=FALSE}
filtered_seurat <- harmony_integration(filtered_seurat, dims = 1:20)
# harmony converged after 2 iterations
```

# Save integrated seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "test", "cortex_integrated_seurat.rds"))
```

# Integration with covariate
```{r}
# change my function to be able to add covariates 
harmony_integration2 <- function(seurat_object, dims, covariates){
  seurat_object <- RunHarmony(seurat_object,
                              group.by.vars = "sample",
                              reduction = "pca",
                              dims.use = dims,
                              assay.use = "RNA",
                              vars_use = covariates)
  # Here we use all PCs computed from Harmony for UMAP calculation -----
  seurat_object <- RunUMAP(seurat_object, dims = dims, reduction = "harmony", reduction.name = "umap_harmony")
  return(seurat_object)
}

filtered_seurat <- harmony_integration2(filtered_seurat, dims = 1:20, covariates = "dataset")
# harmony converged after 2 iterations
saveRDS(filtered_seurat, file = here("data", "test", "cortex_int_covariate_seurat.rds"))
```


# Clustering
* Since I am using harmony for integration, I will have to ensure to use the RNA assay and the harmony reduction for everything downstream.
* I am using Leiden for clustering in my find_clusters function. 
```{r warning=FALSE}
# check active assay (needs to be RNA)
filtered_seurat@active.assay

# vector of desired resolutions
resolutions <- c(0.4, 0.6, 0.8, 1.0, 1.2, 1.3, 1.4, 1.6, 1.8)

# clustering of resolutions (answer no to miniconda installation)
filtered_seurat <- find_clusters(filtered_seurat,
  dims = 1:20,
  reduction = "harmony",
  resolutions = resolutions
)
```

# Save clustered seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "test", "cortex_clustered_seurat.rds"))
```

# Determine resolution
```{r}
# Plot resolutions between 0.1 and 0.9
tree <- clustree(filtered_seurat@meta.data, prefix = "RNA_snn_res.0.")

# Resolutions between 1 and 1.9
tree2 <- clustree(filtered_seurat@meta.data, prefix = "RNA_snn_res.1.")

#pdf(file = here("results", "intermediate_outputs", "", "clustree_umap.pdf"))
tree
tree2

# plotting UMAP for most stable resolution (1.5)
DimPlot(filtered_seurat,
  reduction = "umap_harmony",
  group.by = "RNA_snn_res.1.5"
)

DimPlot(filtered_seurat,
  reduction = "umap_harmony",
  group.by = "dataset",
  shuffle = TRUE
)

DimPlot(filtered_seurat,
  reduction = "umap_harmony",
  group.by = "dataset",
  shuffle = FALSE
)

table(filtered_seurat@meta.data$region)
#dev.off()
```
```{r}
# Plot resolutions between 0.1 and 0.9
tree <- clustree(filtered_seurat@meta.data, prefix = "RNA_snn_res.0.")

# Resolutions between 1 and 1.9
tree2 <- clustree(filtered_seurat@meta.data, prefix = "RNA_snn_res.1.")

#pdf(file = here("results", "intermediate_outputs", "", "clustree_umap.pdf"))
tree
tree2
#dev.off()
```
# UMAP at 0.8
```{r}
DimPlot(filtered_seurat,
  reduction = "umap_harmony",
  group.by = "RNA_snn_res.0.4",
  label = TRUE
)

DimPlot(filtered_seurat,
  reduction = "umap_harmony",
  group.by = "dataset",
  shuffle = TRUE
)

DimPlot(filtered_seurat,
  reduction = "umap_harmony",
  group.by = "dataset",
  shuffle = FALSE
)
```

# Find marker genes
```{r}
# Set the identity as Leiden with resolution
filtered_seurat <- SetIdent(filtered_seurat, value = "RNA_snn_res.0.8")

# Find marker genes
all_marker_genes2 <- FindAllMarkers(filtered_seurat,
  log2FC.threshold = 0.2,
  test.use = "wilcox",
  min.pct = 0.1,
  only.pos = TRUE,
  max.cells.per.ident = 50,
  assay = "RNA"
)

# Save marker genes
saveRDS(all_marker_genes2, here(
  "data",
  "test",
  "cortex_all_marker_genes.rds"
))

# Look at the top 25 up-regulated genes
top25 <- all_marker_genes2 %>%
  group_by(cluster) %>%
  top_n(-25, p_val_adj)

dim(top25) # 1142 7
```

# Feature Plots
```{r}
FeaturePlot(filtered_seurat, features = c("PDGFRB"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("DOCK8", "APBB1IP", "ARHGAP15"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("SNAP25", "RBFOX1"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("SLC7A11", "GFAP"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("ST18", "MBP"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("TNR", "PCDH15"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("FLT1", "VWF"), label = TRUE)
```

# Assign cell types
```{r}
filtered_seurat <- RenameIdents(filtered_seurat,
  `1`  = "Oligodendrocytes",
  `2`  = "Oligodendrocytes",
  `3`  = "Endothelial cells",
  `4`  = "Endothelial cells",
  `5`  = "Pericytes",
  `6`  = "Pericytes",
  `7`  = "Astrocytes",
  `8`  = "Astrocytes",
  `9`  = "Microglia",
  `10` = "Neurons",
  `11` = "Astrocytes",
  `12` = "Microglia_2",
  `13` = "Neurons",
  `14` = "Endothelial cells",
  `15` = "Neurons",
  `16` = "Excitatory Neurons",
  `17` = "Microglia",
  `18` = "OPCs",
  `19` = "Endothelial cells",
  `20` = "OPCs",
  `21` = "Neurons",
  `22` = "Neurons",
  `23` = "Glia - Unspecified",
  `24` = "Neurons",
  `25` = "Pericytes",
  `26` = "Neurons",
  `27` = "Neurons",
  `28` = "Epithelial cells",
  `29` = "Astrocytes",
  `30` = "Neurons",
  `31` = "Neurons"
)
```

# Subclustering of Neuron clusters
```{r}
# Subcluster the neuron clusters
subclustered_seurat <- FindSubCluster(filtered_seurat,
  cluster = "Neurons",
  graph.name = "RNA_snn",
  resolution = 0.1
)

# Set identity to resolution with sub clusters
subclustered_seurat <- SetIdent(subclustered_seurat, value = "sub.cluster")

# Sub-clusters for marker identification
idents <- c("Neurons_0", "Neurons_1", "Neurons_2", "Neurons_3", "Neurons_4", "Neurons_5", "Neurons_6", "Neurons_7", "Neurons_8", "Neurons_9", "Neurons_10", "Neurons_11")

# Find markers for sub-clusters
neurons_markers <- find_markers(subclustered_seurat,
  resolution = "sub.cluster",
  identities = idents,
  value = 5
)

dim(neurons_markers) # 60 6

DimPlot(subclustered_seurat, label = TRUE)

# Assign new cell types
filtered_seurat_test <- RenameIdents(subclustered_seurat,
  "Neurons_0"  = "Inhibitory Neurons",
  "Neurons_1"  = "Excitatory Neurons",
  "Neurons_2"  = "Inhibitory Neurons",
  "Neurons_3"  = "Excitatory Neurons",
  "Neurons_4"  = "Inhibitory Neurons",
  "Neurons_5"  = "Inhibitory Neurons",
  "Neurons_6"  = "Inhibitory Neurons",
  "Neurons_7"  = "Excitatory Neurons",
  "Neurons_8"  = "Excitatory Neurons",
  "Neurons_9"  = "Excitatory Neurons",
  "Neurons_10" = "Inhibitory Neurons",
  "Neurons_11" = "OPcs")

DimPlot(filtered_seurat_test)
```
# save object
```{r}
saveRDS(filtered_seurat_test, file = here("data", "test", "cortex_processed_seurat.rds"))
```

