---
title: "07_testing"
author: "Tabea M. Soelter"
date: '2023-06-01'
output: html_document
---
### Integration and CCC inference of 8 hip samples (VINE)
* Crude code to test whether 8 hip samples are under-powered for LIANA and NN. 
  * Previous analysis yielded no NN outputs, so testing just 8 samples here. 
* This includes single cell QC, integration, and cell type assignment as well as CCC inference.

__Analysis__:
# Load packages
```{r}
# Time tracking
ptm <- proc.time()

# Setting seed
set.seed(42)

# Load packages
suppressPackageStartupMessages({
  library(Seurat)
  library(tidyverse)
  library(harmony)
  library(clustree)
  library(stringr)
  library(scCustomize)
  library(here)
  library(styler)
  library(lintr)
})

source(here("src", "functions_CCCinAD_tms.R"))
```

# Create seurat object
```{r}
merged_seurat <- make_seurat_object(here("data", "CellRangerCounts", "test", "hip8"))
```

# Calculate QC metrics
```{r}
merged_seurat <- calculate_qc(merged_seurat)
```

# Format metadata
```{r}
metadata <- format_metadata(merged_seurat)
dim(metadata) # 143825 7

# Add metadata to merged_seurat
merged_seurat@meta.data <- metadata
```

# Plot QC metrics
```{r warning=FALSE}
#pdf(file = here("results", "intermediate_outputs", "", "qc_plots.pdf"))
plot_qc(metadata)
#dev.off()
```

# Filtering 
* Performing cell-level and gene-level filtering on the data. 
* Cell-level filtering is performed based on the plotted QC metrics (above).
* Gene-level filtering aims at removing count values = 0, since they can skew average gene expression
measurements.
  * Discussed in the following review: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02601-5)
```{r}
# Cell-level filtering
sub_seurat <- subset(merged_seurat,
  subset =
    nGene > 300 &
      nGene < 3000 &
      mitoRatio < 0.10 &
      log10GenesPerUMI > 0.80
)

# Gene-level filtering
counts <- GetAssayData(sub_seurat, slot = "counts")

nonzero <- counts > 0

keep_genes <- rowSums(nonzero) >= 10

filtered_counts <- counts[keep_genes, ]

filtered_seurat <- CreateSeuratObject(filtered_counts,
  meta.data = sub_seurat@meta.data
)
```

# Save filtered seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "test", "hip8_filtered_seurat.rds"))
```

# Plot QC metrics after filtering
```{r warning=FALSE}
metadata <- filtered_seurat@meta.data
dim(metadata) # 117801 10

# plot and save post-filter QC outputs
#pdf(file = here(
#  "results",
#  "intermediate_outputs",
#  "",
#  "qc_plots_filtered.pdf"
#))
plot_qc(metadata)
#dev.off()
```

# Determine cell cycle effects
```{r warning=FALSE}
#pdf(file = here("results", "intermediate_outputs", "", "cellcycle_pca.pdf"))
filtered_seurat <- cell_cylce_effects(filtered_seurat)
#dev.off()
```
# Normalization and integration
* We are using harmony as it has been shown to remove technical variation while preserving biological variation.
* We will have to provide the PCs from the elbow plot above (1:30)
```{r warning=FALSE}
filtered_seurat <- harmony_integration(filtered_seurat, dims = 1:30)
# harmony converged after 6 iterations
```

# Save integrated seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "test", "hip8_integrated_seurat.rds"))
```

# Clustering
* Since I am using harmony for integration, I will have to ensure to use the RNA assay and the harmony reduction for everything downstream.
* I am using Leiden for clustering in my find_clusters function. 
```{r warning=FALSE}
# check active assay (needs to be RNA)
filtered_seurat@active.assay

# vector of desired resolutions
resolutions <- c(0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.2, 1.3, 1.4, 1.5, 1.6)

# clustering of resolutions (answer no to miniconda installation)
filtered_seurat <- find_clusters(filtered_seurat,
  dims = 1:30,
  reduction = "harmony",
  resolutions = resolutions
)
```

# Save clustered seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "test", "hip8_clustered_seurat.rds"))
```

# Determine resolution
```{r}
# Plot resolutions between 0.1 and 0.9
tree <- clustree(filtered_seurat@meta.data, prefix = "RNA_snn_res.0.")

# Resolutions between 1 and 1.9
tree2 <- clustree(filtered_seurat@meta.data, prefix = "RNA_snn_res.1.")

#pdf(file = here("results", "intermediate_outputs", "", "clustree_umap.pdf"))
tree
tree2

# plotting UMAP for most stable resolution (0.6)
DimPlot(filtered_seurat,
  reduction = "umap_harmony",
  group.by = "RNA_snn_res.0.6"
)

DimPlot(vine_clustered_seurat,
  reduction = "umap_harmony",
  group.by = "RNA_snn_res.1.5"
)

#dev.off()
```

# Find marker genes
```{r}
# Set the identity as Leiden with resolution 0.6
filtered_seurat <- SetIdent(filtered_seurat, value = "RNA_snn_res.0.6")

# Find marker genes
all_marker_genes <- FindAllMarkers(filtered_seurat,
  log2FC.threshold = 0.2,
  test.use = "wilcox",
  min.pct = 0.1,
  only.pos = TRUE,
  max.cells.per.ident = 50,
  assay = "RNA"
)

# Save marker genes
saveRDS(all_marker_genes, here(
  "data",
  "test",
  "hip8_all_marker_genes.rds"
))

# Look at the top 25 up-regulated genes
top25 <- all_marker_genes %>%
  group_by(cluster) %>%
  top_n(-25, p_val_adj)

dim(top25) # 882 7
```

# Feature Plots
```{r}
FeaturePlot(filtered_seurat, features = c("PDGFRB"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("DOCK8", "APBB1IP", "ARHGAP15"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("SNAP25", "RBFOX1"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("SLC7A11", "GFAP"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("ST18", "MBP", "IL1RAPL1"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("TNR", "PCDH15", "OPCML"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("FLT1", "VWF"), label = TRUE)

# find marker genes across clusters of interest
#top_markers <- find_markers(vine_clustered_seurat,
#  resolution = "RNA_snn_res.1.5",
#  identities = "42",
#  value = 25
#)
```

# Assign cell types
```{r}
filtered_seurat <- RenameIdents(filtered_seurat,
  `1`  = "Oligodendrocytes",
  `2`  = "Endothelial cells",
  `3`  = "Pericytes",
  `4`  = "Astrocytes",
  `5`  = "Endothelial cells",
  `6`  = "Endothelial cells",
  `7`  = "Pericytes",
  `8`  = "Endothelial cells",
  `9`  = "Astrocytes",
  `10` = "Pericytes",
  `11` = "Microglia",
  `12` = "Microglia",
  `13` = "OPCs",
  `14` = "Endothelial cells",
  `15` = "Oligodendrocytes",
  `16` = "Oligodendrocytes",
  `17` = "Neurons",
  `18` = "Excitatory Neurons (Cilia)",
  `19` = "Astrocytes",
  `20` = "Epithelial cells",
  `21` = "Microglia",
  `22` = "Endothelial cells"
)
```

# Subclustering of cluster 18
```{r}
# Subcluster the neuron clusters
filtered_seurat <- FindSubCluster(filtered_seurat,
  cluster = "Neurons",
  graph.name = "RNA_snn",
  resolution = 0.2
)

# Set identity to resolution with sub clusters
filtered_seurat <- SetIdent(filtered_seurat, value = "sub.cluster")

# Sub-clusters for marker identification
idents <- c("Neurons_0", "Neurons_1", "Neurons_2", "Neurons_3")

# Find markers for sub-clusters
neurons_markers <- find_markers(filtered_seurat,
  resolution = "sub.cluster",
  identities = idents,
  value = 5
)

dim(neurons_markers) # 20 6

DimPlot(filtered_seurat, label = TRUE)

# Assign new cell types
filtered_seurat_test <- RenameIdents(filtered_seurat,
  "Neurons_0" = "Excitatory Neurons",
  "Neurons_1" = "Inhibitory Neurons",
  "Neurons_2" = "Excitatory Neurons",
  "Neurons_3" = "Excitatory Neurons"
)

DimPlot(filtered_seurat_test)
```

# save object
```{r}
saveRDS(filtered_seurat_test, file = here("data", "test", "hip8_processed_seurat.rds"))
```

### CCC

# split object by brain region 
```{r}
split_object <- SplitObject(filtered_seurat_test, split.by = "orig.ident")
```

```{r}
split_object_list <- lapply(filtered_seurat_test, split_objects)

liana_ligands <- lapply(split_object,
  liana_prioritization,
  sender = c(
    "Astrocytes",
    "Oligodendrocytes",
    "Microglia",
    "OPCs"
  ),
  receiver = c(
    "Excitatory Neurons",
    "Inhibitory Neurons"
  )
)
```

```{r}
# save list of ligands
saveRDS(liana_ligands, file = here(
  "data",
  "test",
  "hip8_liana_ligands_unfiltered.rds"
))
```

# Load and filter NicheNet prior
* The GRN and PPI matrices provided by the Saeys lab were downloaded here: https://zenodo.org/record/3260758#.ZEgJROxuf0o
```{r}
# load the ligand target matrix
ligand_target_matrix <- readRDS(here(
  "data",
  "ccc",
  "nichenet_prior",
  "ligand_target_matrix.rds"
))

# load and filter the ligand receptor network as recommended by developers
lr_network <- readRDS(here(
  "data",
  "ccc",
  "nichenet_prior",
  "lr_network.rds"
)) %>%
  mutate(bonafide = !database %in% c("ppi_prediction", "ppi_prediction_go")) %>%
  dplyr::rename(ligand = from, receptor = to) %>%
  distinct(ligand, receptor, bonafide)

# set organism
organism <- "human"
```

# Filter LIANA ligands
* I am filtering ligands identified using LIANA by what is within NicheNet's prior.
```{r}
# filter ligands of each dataset and condition
liana_ligands_filt <- lapply(
  liana_ligands,
  filter_ligands
)

# number of unique ligands: 48 41

# unlist and add objects to global environment
list2env(liana_ligands_filt, globalenv())
``` 

# Collapse and save ligands
```{r}
ligands <- intersect(AD, CTRL) %>% unique()
length(ligands) # 40

# save ligands
saveRDS(ligands, file = here(
  "data",
  "test",
  "hip8_liana_ligands.rds"
))
```

##### NICHENET

```{r}
object_list <- tibble::lst(filtered_seurat_test)

# Making cell-type aggregate object list and save celltype aggregate UMAPs
ca_object_list <- prep_nichenet(object_list,
  file_path = "data/test/"
)

# unlist and add objects to global environment for downstream use
list2env(ca_object_list, globalenv())
```

# set niches
```{r}
## Set niches
user_niches <- list(
  "AD_niche" = list(
    "sender" = c(
      "Astrocytes_AD",
      "Oligodendrocytes_AD",
      "OPCs_AD",
      "Microglia_AD"
    ),
    "receiver" = c("Inhibitory Neurons_AD")
  ),
  "CTRL_niche" = list(
    "sender" = c(
      "Astrocytes_CTRL",
      "Oligodendrocytes_CTRL",
      "OPCs_CTRL",
      "Microglia_CTRL"
    ),
    "receiver" = c("Inhibitory Neurons_CTRL")
  )
)
```

# Differential NicheNet - Excitatory Neurons
* Ligand-receptor and target identification
* The resulting object ("output") gets saved to results/intermediate_outputs/{dataset}/ccc/
* Output include all identified ligands, receptors, and target genes across conditions and between sender and receiver cell types.
* I also use ligands identified by liana in 01_ligand_prioritization_liana.Rmd to filter NicheNet inputs to only include ligands identified by 2 methods.
```{r}
# hip8
nichenet_wrapper(filtered_seurat_test,
  user_niches = user_niches,
  lr_network = lr_network,
  ligands = ligands,
  file_path = "data/test/",
  file_name = "ex_neurons_hip8_output.rds"
)
```






