---
title: "05_integration_all"
author: "Tabea M. Soelter"
date: '2023-06-02'
output: html_document
---
### Integration of all data (vine, vine_cortex, and geo)
* Crude code, not cleaned or styled

__Analysis__:
# Load packages
```{r}
# Time tracking
ptm <- proc.time()

# Setting seed
set.seed(42)

# Load packages
suppressPackageStartupMessages({
  library(Seurat)
  library(tidyverse)
  library(harmony)
  library(clustree)
  library(stringr)
  library(scCustomize)
  library(here)
  library(styler)
  library(lintr)
})

source(here("src", "functions_CCCinAD_tms.R"))
```

# Editing make seurat object function
* I want to add brain region information to the metadata
```{r}
make_seurat_object2 <- function(path){
  counts_list <- list.dirs(path, 
                           full.names = TRUE, 
                           recursive = FALSE)
  object_list <- vector('list')
  for (i in counts_list){
    counts <- Read10X(i)
    print(i)
    sample_name <- basename(i)
    sample_name <- gsub("[[:punct:]]", "", sample_name)
    seurat_object <- CreateSeuratObject(counts = counts, project = sample_name, min.features = 200)
    if (str_sub(sample_name, - 2, - 1) == "AD") {
      seurat_object$orig.ident <- "AD"
    } else {
      seurat_object$orig.ident <- "CTRL"
    }
    if (str_sub(sample_name, - 5, - 3) == "hip") {
      seurat_object$region <- "hippocampus"
    } else {
      seurat_object$region <- "cortex"
    }
    object <- seurat_object
    object_list[[i]] <- object
  } 
  print("Making diseased Seurat Object")
  AD_list <- object_list[grepl("AD", names(object_list))]
  for (i in names(AD_list)) {
    sample_name <- basename(i)
    sample_name <- gsub("[[:punct:]]", "", sample_name)
    AD_list[[i]] <- RenameCells(AD_list[[i]],
                                add.cell.id = sample_name)
  }
  diseased <- Merge_Seurat_List(AD_list)
  
  print("Making control Seurat Object")
  CTRL_list <- object_list[grepl("CTRL", names(object_list))]
  for (i in names(CTRL_list)) {
    sample_name <- basename(i)
    sample_name <- gsub("[[:punct:]]", "", sample_name)
    CTRL_list[[i]] <- RenameCells(CTRL_list[[i]],
                                  add.cell.id = sample_name)
  }
  control <- Merge_Seurat_List(CTRL_list)
  
  print("Making merged Seurat Object")
  merged_seurat <- merge(x = control,
                         y = diseased,
                         add.cell.id = c("CTRL", "AD"))
  return(merged_seurat)
}
```

# Create seurat object
```{r}
merged_seurat <- make_seurat_object2(here("data", "CellRangerCounts", "test", "all"))
```

# Calculate QC metrics
```{r}
merged_seurat <- calculate_qc(merged_seurat)
```

# Format metadata
```{r}
metadata <- format_metadata(merged_seurat)
dim(metadata) # 833943 8

# Add metadata to merged_seurat
merged_seurat@meta.data <- metadata
```

# Plot QC metrics
```{r warning=FALSE}
#pdf(file = here("results", "intermediate_outputs", "", "qc_plots.pdf"))
plot_qc(metadata)
#dev.off()
```

# Filtering 
* Performing cell-level and gene-level filtering on the data. 
* Cell-level filtering is performed based on the plotted QC metrics (above).
* Gene-level filtering aims at removing count values = 0, since they can skew average gene expression
measurements.
  * Discussed in the following review: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02601-5)
```{r}
# Cell-level filtering
sub_seurat <- subset(merged_seurat,
  subset =
    nGene > 300 &
      nGene < 10000 &
      mitoRatio < 0.20 &
      log10GenesPerUMI > 0.80
)

# Gene-level filtering
counts <- GetAssayData(sub_seurat, slot = "counts")

nonzero <- counts > 0

keep_genes <- rowSums(nonzero) >= 10

filtered_counts <- counts[keep_genes, ]

filtered_seurat <- CreateSeuratObject(filtered_counts,
  meta.data = sub_seurat@meta.data
)
```

# Save filtered seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "test", "all_filtered_seurat.rds"))
```

# Plot QC metrics after filtering
```{r warning=FALSE}
metadata <- filtered_seurat@meta.data
dim(metadata) # 688510 11

# plot and save post-filter QC outputs
#pdf(file = here(
#  "results",
#  "intermediate_outputs",
#  "",
#  "qc_plots_filtered.pdf"
#))
plot_qc(metadata)
#dev.off()
```

# Determine cell cycle effects
```{r warning=FALSE}
#pdf(file = here("results", "intermediate_outputs", "", "cellcycle_pca.pdf"))
filtered_seurat <- cell_cylce_effects(filtered_seurat)
#dev.off()
```

# Normalization and integration
* We are using harmony as it has been shown to remove technical variation while preserving biological variation.
* We will have to provide the PCs from the elbow plot above (1:20)
```{r warning=FALSE}
filtered_seurat <- harmony_integration(filtered_seurat, dims = 1:20)
# harmony converged after 2 iterations
```

# Save integrated seurat object
```{r}
saveRDS(filtered_seurat, file = here("data", "test", "all_integrated_seurat.rds"))
```

# Confirm integration
```{r}
DimPlot(filtered_seurat, group.by = "region")
DimPlot(filtered_seurat, group.by = "region", shuffle = TRUE)
```

# Feature Plots
```{r}
FeaturePlot(filtered_seurat, features = c("PDGFRB"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("DOCK8", "APBB1IP", "ARHGAP15"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("SNAP25", "RBFOX1"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("SLC7A11", "GFAP"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("ST18", "MBP", "IL1RAPL1"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("TNR", "PCDH15", "OPCML"), label = TRUE)

FeaturePlot(filtered_seurat, features = c("FLT1", "VWF"), label = TRUE)

# find marker genes across clusters of interest
#top_markers <- find_markers(vine_clustered_seurat,
#  resolution = "RNA_snn_res.1.5",
#  identities = "42",
#  value = 25
#)
```

