---
title: "01_differential_ccc_multinichenet"
author: "Tabea M. Soelter"
date: '2023-07-19'
output: html_document
---
**Differential Cell-Cell Communication Analysis using MultiNicheNet**

__Goal__: For every dataset, perform MultiNicheNet analyses between AD and CTRL to identify ligands, receptors, and potential downstream target genes between glia (Astrocytes, Oligodendrocytes, Microglia, and OPCs - senders) and Excitatory/Inhibitory Neurons (receivers).
  
__Reproducibility__: 
* GitHub: lasseignelab/230313_TS_CCCinHumanAD
* Docker: tsoelter/rstudio_ccc_ad
    * Version: 1.0.3 
* HPC: Yes
  * Resources: long partition (150hrs), 10 CPUs, 65GB per CPU

__Data__:
*Seurat objects*
* Name: geo_processed_seurat.rds, gse_processed_seurat.rds, sadick_processed_seurat.rds
* Location: data/seurat_preprocessing/
*Patient metadata*
* Name: morabito_metadata.csv, lau_metadata.csv, sadick_metadata.csv
* Location: data/dataset_metadata/
 
__Analysis Plan__:
* Load necessary packages 
* Load data
  * Seurat objects
  * Patient metadata
* Load and filter NicheNet v2 prior
* Prepare for NicheNet
  * Determine cut-offs and comparisons
* Differential NicheNet using MultiNicheNet
  * Wrapper function
* Save outputs
  
# Load packages
```{r}
# Time tracking
ptm <- proc.time()

# Setting seed
set.seed(42)

# Setting locale
options(encoding = "UTF-8")

# Load packages
suppressPackageStartupMessages({
  library(nichenetr)
  library(multinichenetr)
  library(SingleCellExperiment)
  library(tidyverse)
  library(tibble)
  library(Seurat)
  library(here)
  library(styler)
  library(lintr)
})

source(here("src", "functions_CCCinAD_tms.R"))
```

# NicheNet-v2 prior
* downloaded here: https://zenodo.org/record/7074291
```{r}
lr_network <- readRDS(here(
  "data",
  "ccc",
  "nichenet_v2_prior",
  "lr_network_human_21122021.rds"
)) %>%
  dplyr::rename(ligand = from, receptor = to) %>%
  distinct(ligand, receptor) %>%
  mutate(ligand = make.names(ligand), receptor = make.names(receptor))

ligand_target_matrix <- readRDS(here(
  "data",
  "ccc",
  "nichenet_v2_prior",
  "ligand_target_matrix_nsga2r_final.rds"
))

# format row and column names
colnames(ligand_target_matrix) <- colnames(ligand_target_matrix) %>%
  make.names()
rownames(ligand_target_matrix) <- rownames(ligand_target_matrix) %>%
  make.names()
```

# Load data
```{r}
geo_object <- readRDS(here(
  "data",
  "seurat_preprocessing",
  "geo_processed_seurat.rds"
))

gse_object <- readRDS(here(
  "data",
  "seurat_preprocessing",
  "gse_processed_seurat.rds"
))

sadick_object <- readRDS(here(
  "data",
  "seurat_preprocessing",
  "sadick_processed_seurat.rds"
))
```

# Add patient metadata
* We perviously determined the need to include sex as a covariate for geo and gse during pseudobulking and differential expression analysis. Sadick did not show this trend, therefore we are only including it in the geo and gse datasets.
```{r}
# GEO
morabito_metadata <- read_csv(here(
  "data",
  "dataset_metadata",
  "morabito_metadata.csv"
))

meta <- geo_object@meta.data

# Filter patient metadata
morabito_filt <- morabito_metadata %>%
  select(Sample, Sex)

# make column names lower case
colnames(morabito_filt) <- tolower(colnames(morabito_filt))

# remove sample not included in the data
morabito_filt <- morabito_filt[-20, ]

# add metdata to seurat metadata
meta_new <- left_join(meta, morabito_filt, by = "sample") %>%
  column_to_rownames("cells")

geo_object@meta.data <- meta_new


# GSE
lau_metadata <- read_csv(here(
  "data",
  "dataset_metadata",
  "lau_metadata.csv"
))

meta <- gse_object@meta.data

# Filter patient metadata
lau_filt <- lau_metadata %>%
  select(SAMPLE, SEX)

colnames(lau_filt) <- tolower(colnames(lau_filt))

# Add metdata to seurat metadata
meta_new <- left_join(meta, lau_filt, by = "sample") %>%
  column_to_rownames("cells")

gse_object@meta.data <- meta_new
```

# Convert seurat objects
```{r}
# GEO
geo <- Seurat::as.SingleCellExperiment(geo_object, assay = "RNA")
geo <- alias_to_symbol_SCE(geo, "human") %>% makenames_SCE()

# GSE
gse <- Seurat::as.SingleCellExperiment(gse_object, assay = "RNA")
gse <- alias_to_symbol_SCE(gse, "human") %>% makenames_SCE()

# Sadick
sadick <- Seurat::as.SingleCellExperiment(sadick_object, assay = "RNA")
sadick <- alias_to_symbol_SCE(sadick, "human") %>% makenames_SCE()
```

# Create object list
```{r}
object_list <- lst(geo, gse, sadick)
```

# Make names syntactically valid
```{r}
object_list <- lapply(object_list, make_names_valid)

list2env(object_list, globalenv())
```

# Set variables for wrapper
```{r}
# sample IDs
sample_id <- "sample"

# condition information
group_id <- "orig.ident"

# cell type information
celltype_id <- "ident"

# Other DE variables
covariates <- "sex"
batches <- NA

# Since we have small sample numbers, we do not use adjusted p-values
empirical_pval <- FALSE
p_val_adj <- FALSE

# Number of cores of the HPC session
cores_system <- 10
```

# Prioritization weights
* These weights are recommended by the developers of MultiNicheNet. There are more detailed descriptions in the following step-by-step vignette:
  * https://github.com/saeyslab/multinichenetr/blob/main/vignettes/pairwise_analysis_MISC.md 
```{r}
prioritizing_weights_DE <- c(
  "de_ligand" = 1,
  "de_receptor" = 1
)

prioritizing_weights_activity <- c(
  "activity_scaled" = 2
)

prioritizing_weights_expression_specificity <- c(
  "exprs_ligand" = 2,
  "exprs_receptor" = 2
)

prioritizing_weights_expression_sufficiency <- c(
  "frac_exprs_ligand_receptor" = 1
)

prioritizing_weights_relative_abundance <- c(
  "abund_sender" = 0,
  "abund_receiver" = 0
)

# Combine all weights into 1 variable
prioritizing_weights <- c(
  prioritizing_weights_DE,
  prioritizing_weights_activity,
  prioritizing_weights_expression_specificity,
  prioritizing_weights_expression_sufficiency,
  prioritizing_weights_relative_abundance
)
```

# Contrast table
* The format of the contrast table (1st line of code in the following chunk) requires no spacing, therefore ensure that the contrast_oi table has no unnecessary spaces or characters. 
```{r}
contrasts_oi <- c("'AD-CTRL','CTRL-AD'")

contrast_tbl <- tibble(
  contrast = c("AD-CTRL", "CTRL-AD"),
  group = c("AD", "CTRL")
)
```

# CCC identification
* For geo and gse, which will be adjusted for sex
```{r}
# create list with only geo and gse
object_list <- lst(geo, gse)

# path for abundance expression plots
results_path <- "results/intermediate_outputs/"

# this runs for approx 45 minutes
multinichenet_outputs <- multinichenet_wrapper(
  object_list = object_list,
  results_path = results_path,
  celltype_id = celltype_id,
  sample_id = sample_id,
  group_id = group_id,
  lr_network = lr_network,
  batches = batches,
  contrasts_oi = contrasts_oi,
  contrast_tbl = contrast_tbl,
  covariates = covariates,
  empirical_pval = empirical_pval,
  p_val_adj = p_val_adj,
  cores_system = cores_system,
  ligand_target_matrix = ligand_target_matrix,
  prioritizing_weights = prioritizing_weights
)

# unlist geo and gse objects
list2env(multinichenet_outputs, globalenv())
```

# CCC inference for Sadick et al
* Since PCA did not indicate the need for a covariate, we excluded them here
```{r}
covariates <- NA

object_list <- lst(sadick)

multinichenet_outputs <- multinichenet_wrapper(
  object_list = object_list,
  results_path = results_path,
  celltype_id = celltype_id,
  sample_id = sample_id,
  group_id = group_id,
  lr_network = lr_network,
  batches = batches,
  contrasts_oi = contrasts_oi,
  contrast_tbl = contrast_tbl,
  covariates = covariates,
  empirical_pval = empirical_pval,
  p_val_adj = p_val_adj,
  cores_system = cores_system,
  ligand_target_matrix = ligand_target_matrix,
  prioritizing_weights = prioritizing_weights
)

# unlist the sadick object
list2env(multinichenet_outputs, globalenv())
```


# Save outputs
```{r}
saveRDS(geo_multinichenet_output, here(
  "data",
  "ccc",
  "geo_multinichenet_output_final.rds"
))

saveRDS(gse_multinichenet_output, here(
  "data",
  "ccc",
  "gse_multinichenet_output_final.rds"
))

saveRDS(sadick_multinichenet_output, here(
  "data",
  "ccc",
  "sadick_multinichenet_output_final.rds"
))
```

# Session info
```{r}
sessionInfo() # see output below
```
R version 4.1.3 (2022-03-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.6 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
[1] C

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lintr_3.0.2                 styler_1.9.1               
 [3] here_1.0.1                  SeuratObject_4.1.3         
 [5] Seurat_4.3.0.9011           lubridate_1.9.2            
 [7] forcats_1.0.0               stringr_1.5.0              
 [9] dplyr_1.1.2                 purrr_1.0.1                
[11] readr_2.1.4                 tidyr_1.3.0                
[13] tibble_3.2.1                ggplot2_3.4.2              
[15] tidyverse_2.0.0             SingleCellExperiment_1.16.0
[17] SummarizedExperiment_1.24.0 Biobase_2.54.0             
[19] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        
[21] IRanges_2.28.0              S4Vectors_0.32.4           
[23] BiocGenerics_0.40.0         MatrixGenerics_1.6.0       
[25] matrixStats_1.0.0           multinichenetr_1.0.0       
[27] nichenetr_2.0.0            

loaded via a namespace (and not attached):
  [1] rsvd_1.0.5                Hmisc_5.1-0               ica_1.0-3                
  [4] ps_1.7.5                  class_7.3-20              rprojroot_2.0.3          
  [7] foreach_1.5.2             lmtest_0.9-40             crayon_1.5.2             
 [10] rhdf5filters_1.6.0        rbibutils_2.2.13          MASS_7.3-55              
 [13] nlme_3.1-155              backports_1.4.1           sva_3.42.0               
 [16] rlang_1.1.1               XVector_0.34.0            caret_6.0-94             
 [19] ROCR_1.0-11               irlba_2.3.5.1             callr_3.7.3              
 [22] nloptr_2.0.3              limma_3.50.3              scater_1.22.0            
 [25] BiocParallel_1.28.3       rjson_0.2.21              bit64_4.0.5              
 [28] glue_1.6.2                sctransform_0.3.5.9006    processx_3.8.1           
 [31] pbkrtest_0.5.2            parallel_4.1.3            vipor_0.4.5              
 [34] spatstat.sparse_3.0-2     AnnotationDbi_1.56.2      UpSetR_1.4.0             
 [37] muscat_1.8.2              spatstat.geom_3.2-2       tidyselect_1.2.0         
 [40] fitdistrplus_1.1-11       variancePartition_1.24.1  XML_3.99-0.14            
 [43] zoo_1.8-12                ggpubr_0.6.0              xtable_1.8-4             
 [46] magrittr_2.0.3            evaluate_0.21             Rdpack_2.4               
 [49] scuttle_1.4.0             cli_3.6.1                 zlibbioc_1.40.0          
 [52] rstudioapi_0.14           miniUI_0.1.1.1            sp_2.0-0                 
 [55] rpart_4.1.16              aod_1.3.2                 locfdr_1.1-8             
 [58] shiny_1.7.4.1             BiocSingular_1.10.0       xfun_0.39                
 [61] clue_0.3-64               cluster_2.1.2             caTools_1.18.2           
 [64] tidygraph_1.2.3           KEGGREST_1.34.0           ggrepel_0.9.3            
 [67] listenv_0.9.0             Biostrings_2.62.0         png_0.1-8                
 [70] future_1.33.0             ipred_0.9-14              withr_2.5.0              
 [73] bitops_1.0-7              ggforce_0.4.1             plyr_1.8.8               
 [76] hardhat_1.3.0             dqrng_0.3.0               e1071_1.7-13             
 [79] pROC_1.18.2               coda_0.19-4               pillar_1.9.0             
 [82] gplots_3.1.3              GlobalOptions_0.1.2       cachem_1.0.8             
 [85] GetoptLong_1.0.5          DelayedMatrixStats_1.16.0 vctrs_0.6.3              
 [88] ellipsis_0.3.2            generics_0.1.3            lava_1.7.2.1             
 [91] tools_4.1.3               foreign_0.8-82            beeswarm_0.4.0           
 [94] munsell_0.5.0             tweenr_2.0.2              emmeans_1.8.6            
 [97] proxy_0.4-27              DelayedArray_0.20.0       fastmap_1.1.1            
[100] compiler_4.1.3            abind_1.4-5               httpuv_1.6.11            
[103] plotly_4.10.2             GenomeInfoDbData_1.2.7    prodlim_2023.03.31       
[106] gridExtra_2.3             glmmTMB_1.1.7             edgeR_3.36.0             
[109] lattice_0.20-45           ggnewscale_0.4.9          deldir_1.0-9             
[112] visNetwork_2.1.2          utf8_1.2.3                later_1.3.1              
[115] recipes_1.0.6             jsonlite_1.8.7            scales_1.2.1             
[118] ScaledMatrix_1.2.0        carData_3.0-5             pbapply_1.7-2            
[121] sparseMatrixStats_1.6.0   estimability_1.4.1        genefilter_1.76.0        
[124] lazyeval_0.2.2            promises_1.2.0.1          car_3.1-2                
[127] doParallel_1.0.17         R.utils_2.12.2            goftest_1.2-3            
[130] spatstat.utils_3.0-3      reticulate_1.30           checkmate_2.2.0          
[133] cyclocomp_1.1.0           rmarkdown_2.23            cowplot_1.1.1            
[136] blme_1.0-5                statmod_1.5.0             Rtsne_0.16               
[139] uwot_0.1.16               igraph_1.5.0              HDF5Array_1.22.1         
[142] survival_3.3-1            numDeriv_2016.8-1.1       yaml_2.3.7               
[145] htmltools_0.5.5           memoise_2.0.1             locfit_1.5-9.8           
[148] graphlayouts_1.0.0        viridisLite_0.4.2         digest_0.6.33            
[151] RhpcBLASctl_0.23-42       mime_0.12                 RSQLite_2.3.1            
[154] future.apply_1.11.0       remotes_2.4.2             data.table_1.14.8        
[157] blob_1.2.4                R.oo_1.25.0               DiagrammeR_1.0.10        
[160] labeling_0.4.2            splines_4.1.3             Formula_1.2-5            
[163] Rhdf5lib_1.16.0           RCurl_1.98-1.12           broom_1.0.5              
[166] hms_1.1.3                 rhdf5_2.38.1              DropletUtils_1.14.2      
[169] colorspace_2.1-0          base64enc_0.1-3           ggbeeswarm_0.7.2         
[172] shape_1.4.6               nnet_7.3-17               Rcpp_1.0.11              
[175] RANN_2.6.1                mvtnorm_1.2-2             circlize_0.4.15          
[178] fansi_1.0.4               tzdb_0.4.0                parallelly_1.36.0        
[181] ModelMetrics_1.2.2.2      R6_2.5.1                  grid_4.1.3               
[184] factoextra_1.0.7          ggridges_0.5.4            lifecycle_1.0.3          
[187] bluster_1.4.0             ggsignif_0.6.4            minqa_1.2.5              
[190] leiden_0.4.3              Matrix_1.5-4              desc_1.4.2               
[193] RcppAnnoy_0.0.21          RColorBrewer_1.1-3        iterators_1.0.14         
[196] spatstat.explore_3.2-1    TMB_1.9.4                 gower_1.0.1              
[199] R.cache_0.16.0            htmlwidgets_1.6.2         beachmat_2.10.0          
[202] polyclip_1.10-4           shadowtext_0.1.2          timechange_0.2.0         
[205] mgcv_1.8-39               ComplexHeatmap_2.10.0     globals_0.16.2           
[208] htmlTable_2.4.1           patchwork_1.1.2           spatstat.random_3.1-5    
[211] progressr_0.13.0          codetools_0.2-18          metapod_1.2.0            
[214] randomForest_4.7-1.1      gtools_3.9.4              prettyunits_1.1.1        
[217] R.methodsS3_1.8.2         gtable_0.3.3              DBI_1.1.3                
[220] tensor_1.5                httr_1.4.6                KernSmooth_2.23-20       
[223] vroom_1.6.3               stringi_1.7.12            progress_1.2.2           
[226] reshape2_1.4.4            farver_2.1.1              annotate_1.72.0          
[229] viridis_0.6.3             rex_1.2.1                 fdrtool_1.2.17           
[232] timeDate_4022.108         xml2_1.3.3                boot_1.3-28              
[235] BiocNeighbors_1.12.0      lme4_1.1-33               geneplotter_1.72.0       
[238] scattermore_1.2           scran_1.22.1              DESeq2_1.34.0            
[241] bit_4.0.5                 spatstat.data_3.0-1       ggraph_2.1.0             
[244] pkgconfig_2.0.3           lmerTest_3.1-3            rstatix_0.7.2            
[247] knitr_1.43 

# Time tracking
```{r}
fptm <- proc.time() - ptm

# convert to minutes
(fptm[3] / 60)

# Time elapsed: 60 minutes
# Hands-on time however is between 15 minutes!
```

# Reproducibility
```{r}
# styler
style_file(here(
  "src",
  "ccc",
  "01_differential_ccc_multinichenet.Rmd"
))

# lintr
lint(filename = here(
  "src",
  "ccc",
  "01_differential_ccc_multinichenet.Rmd"
))
```
